From 79978d7a3849f09b27bb46108596b6a139b87382 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Fri, 5 Jan 2018 13:21:16 -0500
Subject: [PATCH 028/112] disable LEGACY_PTYS by default

---
 drivers/tty/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/tty/Kconfig b/drivers/tty/Kconfig
index e15cd6b5bb99..930be8d5d81d 100644
--- a/drivers/tty/Kconfig
+++ b/drivers/tty/Kconfig
@@ -122,7 +122,6 @@ config UNIX98_PTYS
 
 config LEGACY_PTYS
 	bool "Legacy (BSD) PTY support"
-	default y
 	help
 	  A pseudo terminal (PTY) is a software device consisting of two
 	  halves: a master and a slave. The slave device behaves identical to
-- 
2.31.0

