From b3c60b20a592cbf6308e5a40984ab53386646d38 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Tue, 4 Jul 2017 01:32:30 -0400
Subject: [PATCH 056/112] mark open_softirq as only used for init

---
 include/linux/interrupt.h | 2 +-
 kernel/softirq.c          | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/linux/interrupt.h b/include/linux/interrupt.h
index bb8ff9083e7d..c486c261a997 100644
--- a/include/linux/interrupt.h
+++ b/include/linux/interrupt.h
@@ -578,7 +578,7 @@ static inline void do_softirq_own_stack(void)
 }
 #endif
 
-extern void open_softirq(int nr, void (*action)(struct softirq_action *));
+extern void __init open_softirq(int nr, void (*action)(struct softirq_action *));
 extern void softirq_init(void);
 extern void __raise_softirq_irqoff(unsigned int nr);
 
diff --git a/kernel/softirq.c b/kernel/softirq.c
index 9d71046ea247..adbdc23edb32 100644
--- a/kernel/softirq.c
+++ b/kernel/softirq.c
@@ -483,7 +483,7 @@ void __raise_softirq_irqoff(unsigned int nr)
 	or_softirq_pending(1UL << nr);
 }
 
-void open_softirq(int nr, void (*action)(struct softirq_action *))
+void __init open_softirq(int nr, void (*action)(struct softirq_action *))
 {
 	softirq_vec[nr].action = action;
 }
-- 
2.31.0

