From 48c7e5608741bd9270a64a5df2cd064042fdbad4 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 20 Aug 2017 15:39:25 -0400
Subject: [PATCH 011/113] enable SLAB_FREELIST_HARDENED by default

---
 init/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/Kconfig b/init/Kconfig
index dcbcb4243316..667d1c6c021b 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1938,6 +1938,7 @@ config SLAB_FREELIST_RANDOM
 config SLAB_FREELIST_HARDENED
 	bool "Harden slab freelist metadata"
 	depends on SLAB || SLUB
+	default y
 	help
 	  Many kernel heap attacks try to target slab cache metadata and
 	  other infrastructure. This options makes minor performance
-- 
2.30.0

