--- gcc/gcc/config/sparc/linux.h.piesparcdef~	2005-03-01 17:18:15 +0100
+++ gcc/gcc/config/sparc/linux.h	2005-03-01 17:25:51 +0100
@@ -48,8 +48,12 @@
    provides part of the support for getting C++ file-scope static
    object constructed before entering `main'.  */
    
+#undef  HARDENED_STARTFILE_SPEC
 #undef  STARTFILE_SPEC
 #ifdef HAVE_LD_PIE
+#define HARDENED_STARTFILE_SPEC \
+  "%{!shared: %{pg|p:gcrt1.o%s;nopie:crt1.o%s;:Scrt1.o%s}}\
+   crti.o%s %{static:crtbeginT.o%s;nopie:crtbegin.o%s;:crtbeginS.o%s}"
 #define STARTFILE_SPEC \
   "%{!shared: %{pg|p:gcrt1.o%s;pie:Scrt1.o%s;:crt1.o%s}}\
    crti.o%s %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
@@ -65,8 +69,12 @@
    object constructed before entering `main', followed by a normal
    GNU/Linux "finalizer" file, `crtn.o'.  */
 
+#undef  HARDENED_ENDFILE_SPEC
 #undef  ENDFILE_SPEC
 #ifdef HAVE_LD_PIE
+#define HARDENED_ENDFILE_SPEC \
+  "%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
+   %{static|nopie:crtend.o%s;:crtendS.o%s} crtn.o%s"
 #define ENDFILE_SPEC \
   "%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
    %{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s"
@@ -153,6 +161,17 @@
   "%{V} %{v:%{!V:-V}} %{!Qn:-Qy} %{n} %{T} %{Ym,*} %{Wa,*:%*} -s \
    %{fpic|fPIC|fpie|fPIE|pie:-K PIC} %(asm_cpu) %(asm_relax)"
 
+#ifdef HAVE_LD_PIE
+#if defined(EFAULT_PIE) || defined(EFAULT_PIE_SSP)
+#define HARDENED_ASM_SPEC \
+  "%{V} %{v:%{!V:-V}} %{!Qn:-Qy} %{n} %{T} %{Ym,*} %{Wa,*:%*} -s \
+   %{!D__KERNEL__: %{!fno-PIC:%{!fno-pic:%{!fno-PIE:%{!fno-pie: %{!nopie:-K PIC} }}}} } \
+   %(asm_cpu) %(asm_relax)"
+#undef ASM_SPEC
+#define ASM_SPEC HARDENED_ASM_SPEC
+#endif
+#endif
+
 /* Same as sparc.h */
 #undef DBX_REGISTER_NUMBER
 #define DBX_REGISTER_NUMBER(REGNO) (REGNO)
--- gcc/gcc/config/sparc/linux64.h.piesparcdef~	2005-03-01 17:18:15 +0100
+++ gcc/gcc/config/sparc/linux64.h	2005-03-01 17:24:45 +0100
@@ -71,9 +71,13 @@
    provides part of the support for getting C++ file-scope static
    object constructed before entering `main'.  */
    
+#undef  HARDENED_STARTFILE_SPEC
 #undef  STARTFILE_SPEC
 
 #ifdef HAVE_LD_PIE
+#define HARDENED_STARTFILE_SPEC \
+  "%{!shared:%{pg|p:gcrt1.o%s;static|nopie:crt1.o%s;:Scrt1.o%s}}\
+   crti.o%s %{static:crtbeginT.o%s;nopie:crtbegin.o%s;:crtbeginS.o%s}"
 #define STARTFILE_SPEC \
   "%{!shared:%{pg|p:gcrt1.o%s;pie:Scrt1.o%s;:crt1.o%s}}\
    crti.o%s %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbeginS.o%s}"
@@ -89,8 +93,12 @@
    object constructed before entering `main', followed by a normal
    GNU/Linux "finalizer" file, `crtn.o'.  */
 
+#undef  HARDENED_ENDFILE_SPEC
 #undef  ENDFILE_SPEC
 #ifdef HAVE_LD_PIE
+#define HARDENED_ENDFILE_SPEC \
+  "%{static|nopie:crtend.o%s;:crtendS.o%s} crtn.o%s\
+   %{ffast-math|funsafe-math-optimizations:crtfastmath.o%s}"
 #define ENDFILE_SPEC \
   "%{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s\
    %{ffast-math|funsafe-math-optimizations:crtfastmath.o%s}"
@@ -290,6 +298,24 @@
 %{mlittle-endian:-EL} \
 %(asm_cpu) %(asm_arch) %(asm_relax)"
 
+#ifdef HAVE_LD_PIE
+#if defined(EFAULT_PIE) || defined(EFAULT_PIE_SSP)
+#define HARDENED_ASM_SPEC "\
+%{V} \
+%{v:%{!V:-V}} \
+%{!Qn:-Qy} \
+%{n} \
+%{T} \
+%{Ym,*} \
+%{Wa,*:%*} \
+-s %{!D__KERNEL__: %{!fno-PIC:%{!fno-pic:%{!fno-PIE:%{!fno-pie: %{!nopie:-K PIC} }}}} } \
+%{mlittle-endian:-EL} \
+%(asm_cpu) %(asm_arch) %(asm_relax)"
+#undef ASM_SPEC
+#define ASM_SPEC HARDENED_ASM_SPEC
+#endif
+#endif
+
 /* Same as sparc.h */
 #undef DBX_REGISTER_NUMBER
 #define DBX_REGISTER_NUMBER(REGNO) (REGNO)
