From a96b01e00822404618c4060faf9421ab63544cd6 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sat, 27 May 2017 07:28:10 -0400
Subject: [PATCH 024/113] disable DEVPORT by default

---
 drivers/char/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/char/Kconfig b/drivers/char/Kconfig
index d229a2d0c017..68178c3a25de 100644
--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -391,7 +391,6 @@ config MAX_RAW_DEVS
 config DEVPORT
 	bool "/dev/port character device"
 	depends on ISA || PCI
-	default y
 	help
 	  Say Y here if you want to support the /dev/port device. The /dev/port
 	  device is similar to /dev/mem, but for I/O ports.
-- 
2.30.0

