From 361c8a6666553e88e9446241fa75962184aab482 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 25 Feb 2018 02:15:24 -0500
Subject: [PATCH 037/113] enable AUDIT by default

---
 init/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/Kconfig b/init/Kconfig
index 4f5827e10be3..9b75a4921575 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -419,6 +419,7 @@ config USELIB
 config AUDIT
 	bool "Auditing support"
 	depends on NET
+	default y
 	help
 	  Enable auditing infrastructure that can be used with another
 	  kernel subsystem, such as SELinux (which requires this for
-- 
2.30.0

