From 36de3cb8f8d03632b80a983f4dc63b665cf208b7 Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Fri, 16 Apr 2021 00:41:36 +0200
Subject: [PATCH 112/112] Linux hardened 5.10.30-hardened1

Signed-off-by: Levente Polyak <levente@leventepolyak.net>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 872af26e085a..78ee1329ef60 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 5
 PATCHLEVEL = 10
 SUBLEVEL = 30
-EXTRAVERSION =
+EXTRAVERSION = -hardened1
 NAME = Dare mighty things
 
 # *DOCUMENTATION*
-- 
2.31.1

