From ecbabc26570534a4d356ce3cf26b6c2db4167409 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Tue, 30 May 2017 10:47:23 -0400
Subject: [PATCH 033/102] enable protected_{symlinks,hardlinks} by default

---
 fs/namei.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/fs/namei.c b/fs/namei.c
index 61fdb77a7d58..452d0da59d14 100644
--- a/fs/namei.c
+++ b/fs/namei.c
@@ -918,8 +918,8 @@ static inline void put_link(struct nameidata *nd)
 		path_put(&last->link);
 }
 
-int sysctl_protected_symlinks __read_mostly = 0;
-int sysctl_protected_hardlinks __read_mostly = 0;
+int sysctl_protected_symlinks __read_mostly = 1;
+int sysctl_protected_hardlinks __read_mostly = 1;
 int sysctl_protected_fifos __read_mostly;
 int sysctl_protected_regular __read_mostly;
 
-- 
2.28.0

