From 22cad148f3ea21c0a437f0f44dbac414d650f988 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 25 Feb 2018 02:15:24 -0500
Subject: [PATCH 037/102] enable AUDIT by default

---
 init/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/Kconfig b/init/Kconfig
index e6e14bd4668b..bbb2d461d52e 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -361,6 +361,7 @@ config USELIB
 config AUDIT
 	bool "Auditing support"
 	depends on NET
+	default y
 	help
 	  Enable auditing infrastructure that can be used with another
 	  kernel subsystem, such as SELinux (which requires this for
-- 
2.28.0

