--- gcc/config.gcc.orig	2010-03-03 10:04:06.000000000 -0800
+++ gcc/config.gcc	2010-03-03 10:05:16.000000000 -0800
@@ -2891,8 +2891,9 @@
 
 		case "$with_arch" in
 		"" \
-		| armv[23456] | armv2a | armv3m | armv4t | armv5t \
+		| armv[234567] | armv2a | armv3m | armv4t | armv5t \
 		| armv5te | armv6j |armv6k | armv6z | armv6zk \
+		| armv7-a | armv7-m | armv7-r \
 		| iwmmxt | ep9312)
 			# OK
 			;;
