http://bugs.gentoo.org/135840

reduced backport from newer versions

--- gcc-4.1.2/libjava/classpath/configure
+++ gcc-4.1.2/libjava/classpath/configure
@@ -4252,7 +4252,12 @@
 
 else
 
-	     nativelibdir='${libdir}/${PACKAGE}'
+multi_os_directory=`$CC -print-multi-os-directory`
+case $multi_os_directory in
+	.) toolexeclibdir=${libdir} ;; # Avoid trailing /.
+	*) toolexeclibdir=${libdir}/${multi_os_directory} ;;
+esac
+nativelibdir="${toolexeclibdir}/gcj-"`cat ${srcdir}/../../gcc/BASE-VER`
 
 fi;
 
