From 660199c1d2761a754cff6a4d9dfe7c16b2d5f79f Mon Sep 17 00:00:00 2001
From: Nick Alcock <nick.alcock@oracle.com>
Date: Mon, 9 Dec 2019 16:52:43 +0000
Subject: [PATCH 21/24] fixup! dtrace: function boundary tracing (FBT) core and
 x86 components

---
 arch/x86/kernel/nmi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/kernel/nmi.c b/arch/x86/kernel/nmi.c
index bcbec09b35f5..4e56580a4e4d 100644
--- a/arch/x86/kernel/nmi.c
+++ b/arch/x86/kernel/nmi.c
@@ -513,7 +513,7 @@ dotraplinkage notrace int
 do_nmi(struct pt_regs *regs, long error_code)
 {
 	if (IS_ENABLED(CONFIG_SMP) && cpu_is_offline(smp_processor_id()))
-		return;
+		return 0;
 
 	if (this_cpu_read(nmi_state) != NMI_NOT_RUNNING) {
 		this_cpu_write(nmi_state, NMI_LATCHED);
-- 
2.28.0

