--- gcc-2.95.3/gcc/f/com.h.orig	2005-03-14 02:03:50.000000000 -0800
+++ gcc-2.95.3/gcc/f/com.h	2005-03-14 02:04:01.000000000 -0800
@@ -310,7 +310,7 @@ void ffecom_finish_decl (tree decl, tree
 void ffecom_finish_progunit (void);
 tree ffecom_get_invented_identifier (const char *pattern, const char *text,
 				     int number);
-ffeinfoKindtype ffecom_gfrt_basictype (ffecomGfrt ix);
+ffeinfoBasictype ffecom_gfrt_basictype (ffecomGfrt ix);
 ffeinfoKindtype ffecom_gfrt_kindtype (ffecomGfrt ix);
 void ffecom_init_0 (void);
 void ffecom_init_2 (void);
