From 035b897c45fe08a3090fe7440c0ad1634bb7f62b Mon Sep 17 00:00:00 2001
From: Bernhard40 <32568352+Bernhard40@users.noreply.github.com>
Date: Fri, 6 Oct 2017 10:24:10 +0000
Subject: [PATCH 022/104] disable AIO by default

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index c1b0bda7f403..7bbf4aef6d79 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1553,7 +1553,6 @@ config SHMEM
 
 config AIO
 	bool "Enable AIO support"
-	default y
 	help
 	  This option enables POSIX asynchronous I/O which may by used
 	  by some high performance threaded applications. Disabling
-- 
2.28.0

