From d0d3c29cf8fa2b3bf01ecd86a0eee89513aa0df7 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sat, 27 May 2017 07:28:10 -0400
Subject: [PATCH 024/104] disable DEVPORT by default

---
 drivers/char/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/char/Kconfig b/drivers/char/Kconfig
index 98c3a5d8003e..100cc6e96bcf 100644
--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -390,7 +390,6 @@ config MAX_RAW_DEVS
 config DEVPORT
 	bool "/dev/port character device"
 	depends on ISA || PCI
-	default y
 	help
 	  Say Y here if you want to support the /dev/port device. The /dev/port
 	  device is similar to /dev/mem, but for I/O ports.
-- 
2.28.0

