diff -Nru gcc-3.3.5.old/boehm-gc/configure gcc-3.3.5/boehm-gc/configure
--- gcc-3.3.5.old/boehm-gc/configure	2004-11-03 15:39:28.817915816 -0500
+++ gcc-3.3.5/boehm-gc/configure	2004-11-03 15:39:52.882257480 -0500
@@ -3097,7 +3097,11 @@
   toolexecdir='$(libdir)/gcc-lib/$(target_alias)'
   toolexeclibdir='$(libdir)'
 fi
-toolexeclibdir=$toolexeclibdir/`$CC -print-multi-os-directory`
+multi_os_directory=`$CC -print-multi-os-directory`
+case $multi_os_directory in
+  .) ;; # Avoid trailing /.
+  *) toolexeclibdir=$toolexeclibdir/$multi_os_directory ;;
+esac
 
 
 
diff -Nru gcc-3.3.5.old/boehm-gc/configure.in gcc-3.3.5/boehm-gc/configure.in
--- gcc-3.3.5.old/boehm-gc/configure.in	2004-11-03 15:39:28.839912472 -0500
+++ gcc-3.3.5/boehm-gc/configure.in	2004-11-03 15:39:52.882257480 -0500
@@ -278,7 +278,11 @@
   toolexecdir='$(libdir)/gcc-lib/$(target_alias)'
   toolexeclibdir='$(libdir)'
 fi
-toolexeclibdir=$toolexeclibdir/`$CC -print-multi-os-directory`
+multi_os_directory=`$CC -print-multi-os-directory`
+case $multi_os_directory in
+  .) ;; # Avoid trailing /.
+  *) toolexeclibdir=$toolexeclibdir/$multi_os_directory ;;
+esac
 AC_SUBST(toolexecdir)
 AC_SUBST(toolexeclibdir)
 
diff -Nru gcc-3.3.5.old/libf2c/aclocal.m4 gcc-3.3.5/libf2c/aclocal.m4
--- gcc-3.3.5.old/libf2c/aclocal.m4	2004-11-03 15:39:26.650245352 -0500
+++ gcc-3.3.5/libf2c/aclocal.m4	2004-11-03 15:39:52.883257328 -0500
@@ -212,7 +212,11 @@
     glibcpp_toolexecdir='$(libdir)/gcc-lib/$(target_alias)'
     glibcpp_toolexeclibdir='$(libdir)'
   fi
-  glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/`$CC -print-multi-os-directory`
+  multi_os_directory=`$CC -print-multi-os-directory`
+  case $multi_os_directory in
+    .) ;; # Avoid trailing /.
+    *) glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/$multi_os_directory ;;
+  esac
 fi
 
 AC_SUBST(glibcpp_prefixdir)
diff -Nru gcc-3.3.5.old/libjava/configure gcc-3.3.5/libjava/configure
--- gcc-3.3.5.old/libjava/configure	2004-11-03 15:39:17.788592528 -0500
+++ gcc-3.3.5/libjava/configure	2004-11-03 15:39:52.886256872 -0500
@@ -5775,7 +5775,11 @@
   toolexecdir='$(libdir)/gcc-lib/$(target_alias)'
   toolexecmainlibdir='$(libdir)'
 fi
-toolexeclibdir=$toolexecmainlibdir/`$CC -print-multi-os-directory`
+multi_os_directory=`$CC -print-multi-os-directory`
+case $multi_os_directory in
+  .) toolexeclibdir=$toolexecmainlibdir ;; # Avoid trailing /.
+  *) toolexeclibdir=$toolexecmainlibdir/$multi_os_directory ;;
+esac
 
 
 
diff -Nru gcc-3.3.5.old/libjava/configure.in gcc-3.3.5/libjava/configure.in
--- gcc-3.3.5.old/libjava/configure.in	2004-11-03 15:39:17.839584776 -0500
+++ gcc-3.3.5/libjava/configure.in	2004-11-03 15:39:52.887256720 -0500
@@ -805,7 +805,11 @@
   toolexecdir='$(libdir)/gcc-lib/$(target_alias)'
   toolexecmainlibdir='$(libdir)'
 fi
-toolexeclibdir=$toolexecmainlibdir/`$CC -print-multi-os-directory`
+multi_os_directory=`$CC -print-multi-os-directory`
+case $multi_os_directory in
+  .) toolexeclibdir=$toolexecmainlibdir ;; # Avoid trailing /.
+  *) toolexeclibdir=$toolexecmainlibdir/$multi_os_directory ;;
+esac
 AC_SUBST(toolexecdir)
 AC_SUBST(toolexecmainlibdir)
 AC_SUBST(toolexeclibdir)
diff -Nru gcc-3.3.5.old/libobjc/aclocal.m4 gcc-3.3.5/libobjc/aclocal.m4
--- gcc-3.3.5.old/libobjc/aclocal.m4	2004-11-03 15:39:21.949959904 -0500
+++ gcc-3.3.5/libobjc/aclocal.m4	2004-11-03 15:39:52.887256720 -0500
@@ -212,7 +212,11 @@
     glibcpp_toolexecdir='$(libdir)/gcc-lib/$(target_alias)'
     glibcpp_toolexeclibdir='$(libdir)'
   fi
-  glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/`$CC -print-multi-os-directory`
+  multi_os_directory=`$CC -print-multi-os-directory`
+  case $multi_os_directory in
+  .) ;; # Avoid trailing /.
+  *) glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/$multi_os_directory ;;
+  esac
 fi
 
 AC_SUBST(glibcpp_prefixdir)
diff -Nru gcc-3.3.5.old/libobjc/configure gcc-3.3.5/libobjc/configure
--- gcc-3.3.5.old/libobjc/configure	2004-11-03 15:39:21.954959144 -0500
+++ gcc-3.3.5/libobjc/configure	2004-11-03 15:39:52.889256416 -0500
@@ -1270,7 +1270,11 @@
     glibcpp_toolexecdir='$(libdir)/gcc-lib/$(target_alias)'
     glibcpp_toolexeclibdir='$(libdir)'
   fi
-  glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/`$CC -print-multi-os-directory`
+  multi_os_directory=`$CC -print-multi-os-directory`
+  case $multi_os_directory in
+  .) ;; # Avoid trailing /.
+  *) glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/$multi_os_directory ;;
+  esac
 fi
 
 
diff -Nru gcc-3.3.5.old/libstdc++-v3/acinclude.m4 gcc-3.3.5/libstdc++-v3/acinclude.m4
--- gcc-3.3.5.old/libstdc++-v3/acinclude.m4	2004-11-03 15:39:24.299602704 -0500
+++ gcc-3.3.5/libstdc++-v3/acinclude.m4	2004-11-03 15:39:52.890256264 -0500
@@ -1970,7 +1970,11 @@
     glibcpp_toolexecdir='$(libdir)/gcc-lib/$(target_alias)'
     glibcpp_toolexeclibdir='$(libdir)'
   fi
-  glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/`$CC -print-multi-os-directory`
+  multi_os_directory=`$CC -print-multi-os-directory`
+  case $multi_os_directory in
+  .) ;; # Avoid trailing /.
+  *) glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/$multi_os_directory ;;
+  esac
 fi
 
 AC_MSG_CHECKING([for install location])
diff -Nru gcc-3.3.5.old/libstdc++-v3/aclocal.m4 gcc-3.3.5/libstdc++-v3/aclocal.m4
--- gcc-3.3.5.old/libstdc++-v3/aclocal.m4	2004-11-03 15:39:22.532871288 -0500
+++ gcc-3.3.5/libstdc++-v3/aclocal.m4	2004-11-03 15:39:52.905253984 -0500
@@ -1982,7 +1982,11 @@
     glibcpp_toolexecdir='$(libdir)/gcc-lib/$(target_alias)'
     glibcpp_toolexeclibdir='$(libdir)'
   fi
-  glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/`$CC -print-multi-os-directory`
+  multi_os_directory=`$CC -print-multi-os-directory`
+  case $multi_os_directory in
+  .) ;; # Avoid trailing /.
+  *) glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/$multi_os_directory ;;
+  esac
 fi
 
 AC_MSG_CHECKING([for install location])
diff -Nru gcc-3.3.5.old/libstdc++-v3/configure gcc-3.3.5/libstdc++-v3/configure
--- gcc-3.3.5.old/libstdc++-v3/configure	2004-11-03 15:39:23.246762760 -0500
+++ gcc-3.3.5/libstdc++-v3/configure	2004-11-03 15:41:27.585860344 -0500
@@ -2009,7 +2009,7 @@
 # This must be Linux ELF.
 linux-gnu*)
   case $host_cpu in
-  alpha* | hppa* | i*86 | powerpc* | sparc* | ia64* | sh* )
+  alpha* | hppa* | i*86 | m68* | powerpc* | sparc* | ia64* | sh* )
     lt_cv_deplibs_check_method=pass_all ;;
   *)
     # glibc up to 2.1.1 does not perform some relocations on ARM
@@ -24134,7 +24134,11 @@
     glibcpp_toolexecdir='$(libdir)/gcc-lib/$(target_alias)'
     glibcpp_toolexeclibdir='$(libdir)'
   fi
-  glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/`$CC -print-multi-os-directory`
+  multi_os_directory=`$CC -print-multi-os-directory`
+  case $multi_os_directory in
+  .) ;; # Avoid trailing /.
+  *) glibcpp_toolexeclibdir=$glibcpp_toolexeclibdir/$multi_os_directory ;;
+  esac
 fi
 
 echo $ac_n "checking for install location""... $ac_c" 1>&6
diff -Nru gcc-3.3.5.old/zlib/configure gcc-3.3.5/zlib/configure
--- gcc-3.3.5.old/zlib/configure	2004-11-03 15:39:07.768115872 -0500
+++ gcc-3.3.5/zlib/configure	2004-11-03 15:39:52.919251856 -0500
@@ -2752,6 +2752,9 @@
 fi
 if test "$GCC" = yes && $CC -print-multi-os-directory > /dev/null 2>&1; then
   multiosdir=/`$CC -print-multi-os-directory`
+  case $multiosdir in
+    /.) multiosdir= ;; # Avoid trailing /.
+  esac
 else
   multiosdir=
 fi
diff -Nru gcc-3.3.5.old/zlib/configure.in gcc-3.3.5/zlib/configure.in
--- gcc-3.3.5.old/zlib/configure.in	2004-11-03 15:39:07.794111920 -0500
+++ gcc-3.3.5/zlib/configure.in	2004-11-03 15:39:52.931250032 -0500
@@ -123,6 +123,9 @@
 fi
 if test "$GCC" = yes && $CC -print-multi-os-directory > /dev/null 2>&1; then
   multiosdir=/`$CC -print-multi-os-directory`
+  case $multiosdir in
+    /.) multiosdir= ;; # Avoid trailing /.
+  esac
 else
   multiosdir=
 fi
