Index: gcc/cppexp.c
===================================================================
RCS file: /cvs/gcc/gcc/gcc/cppexp.c,v
retrieving revision 1.136
diff -u -p -r1.136 cppexp.c
--- gcc/cppexp.c	19 Dec 2002 05:18:04 -0000	1.136
+++ gcc/cppexp.c	13 Mar 2003 20:41:04 -0000
@@ -175,7 +175,8 @@ cpp_classify_number (pfile, token)
       str++;
 
       /* Require at least one hex digit to classify it as hex.  */
-      if ((*str == 'x' || *str == 'X') && ISXDIGIT (str[1]))
+      if ((*str == 'x' || *str == 'X')
+	  && (str[1] == '.' || ISXDIGIT (str[1])))
 	{
 	  radix = 16;
 	  str++;
