diff -urpN gcc-3.3.1/ChangeLog gcc-3.3.1.coreutils/ChangeLog
--- gcc-3.3.1/ChangeLog	2003-09-06 22:45:51.000000000 +0200
+++ gcc-3.3.1.coreutils/ChangeLog	2003-09-06 22:56:09.955587560 +0200
@@ -8458,7 +8458,7 @@ Tue Dec 29 15:06:00 1992  Ian Lance Tayl
 	to i386-sco3.2v4.  Set TARGETS and CFLAGS for i386-sco3.2v4.
 	(all-cygnus, native, build-cygnus): Make
 	$(canonhost)-stamp-3stage-done, not $(host)....
-	* test-build.mk (stamp-3stage-compared): Use tail +10c for
+	* test-build.mk (stamp-3stage-compared): Use tail -c +10 for
 	i386-sco3.2v4.  Added else true to if command.
 
 Mon Dec 28 12:08:56 1992  Ken Raeburn  (raeburn@cygnus.com)
diff -urpN gcc-3.3.1/boehm-gc/config.guess gcc-3.3.1.coreutils/boehm-gc/config.guess
--- gcc-3.3.1/boehm-gc/config.guess	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/boehm-gc/config.guess	2003-09-06 22:55:58.514326896 +0200
@@ -340,7 +340,7 @@ EOF
 	echo m68k-sun-sunos${UNAME_RELEASE}
 	exit 0 ;;
     sun*:*:4.2BSD:*)
-	UNAME_RELEASE=`(head -1 /etc/motd | awk '{print substr($5,1,3)}') 2>/dev/null`
+	UNAME_RELEASE=`(head -n 1 /etc/motd | awk '{print substr($5,1,3)}') 2>/dev/null`
 	test "x${UNAME_RELEASE}" = "x" && UNAME_RELEASE=3
 	case "`/bin/arch`" in
 	    sun3)
@@ -509,7 +509,7 @@ EOF
 	fi
 	exit 0 ;;
     *:AIX:*:[45])
-	IBM_CPU_ID=`/usr/sbin/lsdev -C -c processor -S available | head -1 | awk '{ print $1 }'`
+	IBM_CPU_ID=`/usr/sbin/lsdev -C -c processor -S available | head -n 1 | awk '{ print $1 }'`
 	if /usr/sbin/lsattr -El ${IBM_CPU_ID} | grep ' POWER' >/dev/null 2>&1; then
 		IBM_ARCH=rs6000
 	else
diff -urpN gcc-3.3.1/configure gcc-3.3.1.coreutils/configure
--- gcc-3.3.1/configure	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/configure	2003-09-06 22:56:09.983583304 +0200
@@ -820,7 +820,7 @@ if [ -z "${CC}" ] && [ -r Makefile ]; th
 s/\\\n//g
 t loop
 /^CC[ 	]*=/ s/CC[ 	]*=[ 	]*\(.*\)/\1/p' < Makefile > Makefile.cc
-  CC=`tail -1 Makefile.cc`
+  CC=`tail -n 1 Makefile.cc`
   rm -f Makefile.cc
 fi
 
@@ -830,7 +830,7 @@ if [ -z "${CFLAGS}" ] && [ -r Makefile ]
 s/\\\n//g
 t loop
 /^CFLAGS[ 	]*=/ s/CFLAGS[ 	]*=[ 	]*\(.*\)/\1/p' < Makefile > Makefile.cc
-  CFLAGS=`tail -1 Makefile.cc`
+  CFLAGS=`tail -n 1 Makefile.cc`
   rm -f Makefile.cc
 fi
 
@@ -840,7 +840,7 @@ if [ -z "${CXX}" ] && [ -r Makefile ]; t
 s/\\\n//g
 t loop
 /^CXX[ 	]*=/ s/CXX[ 	]*=[ 	]*\(.*\)/\1/p' < Makefile > Makefile.cc
-  CXX=`tail -1 Makefile.cc`
+  CXX=`tail -n 1 Makefile.cc`
   rm -f Makefile.cc
 fi
 
@@ -850,7 +850,7 @@ if [ -z "${CXXFLAGS}" ] && [ -r Makefile
 s/\\\n//g
 t loop
 /^CXXFLAGS[ 	]*=/ s/CXXFLAGS[ 	]*=[ 	]*\(.*\)/\1/p' < Makefile > Makefile.cc
-  CXXFLAGS=`tail -1 Makefile.cc`
+  CXXFLAGS=`tail -n 1 Makefile.cc`
   rm -f Makefile.cc
 fi
 
@@ -931,7 +931,7 @@ s/\\\n//g
 t loop
 /^'"${var}"'[ 	]*=/ s/'"${var}"'[ 	]*=[ 	]*\(.*\)/\1/p' \
 	< Makefile > Makefile.v
-      t=`tail -1 Makefile.v`
+      t=`tail -n 1 Makefile.v`
       if [ -n "${t}" ]; then
 	eval "${var}=\${t}"
       fi
diff -urpN gcc-3.3.1/contrib/test_summary gcc-3.3.1.coreutils/contrib/test_summary
--- gcc-3.3.1/contrib/test_summary	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/contrib/test_summary	2003-09-06 22:56:09.999580872 +0200
@@ -98,7 +98,7 @@ $1 ~ /\/configure$/ {
     srcdir = $1;
     gsub(/\/configure$/, "", srcdir);
     printf "LAST_UPDATED: ";
-    system("tail -1 " srcdir "/LAST_UPDATED");
+    system("tail -n 1 " srcdir "/LAST_UPDATED");
     print "";
 
     $1 = "configure flags:"; configflags = $0;
diff -urpN gcc-3.3.1/gcc/Makefile.in gcc-3.3.1.coreutils/gcc/Makefile.in
--- gcc-3.3.1/gcc/Makefile.in	2003-09-06 22:45:56.000000000 +0200
+++ gcc-3.3.1.coreutils/gcc/Makefile.in	2003-09-06 22:56:10.088567344 +0200
@@ -3484,16 +3484,16 @@ compare compare3 compare4 compare-lean c
 	-rm -f .bad_compare
 	case "$@" in compare | compare-lean ) stage=2 ;; * ) stage=`echo $@ | sed -e 's,^compare\([0-9][0-9]*\).*,\1,'` ;; esac; \
 	for file in *$(objext); do \
-	  tail +16c ./$$file > tmp-foo1; \
-	  tail +16c stage$$stage/$$file > tmp-foo2 \
+	  tail -c +16 ./$$file > tmp-foo1; \
+	  tail -c +16 stage$$stage/$$file > tmp-foo2 \
 	    && (cmp tmp-foo1 tmp-foo2 > /dev/null 2>&1 || echo $$file differs >> .bad_compare) || true; \
 	done
 	case "$@" in compare | compare-lean ) stage=2 ;; * ) stage=`echo $@ | sed -e 's,^compare\([0-9][0-9]*\).*,\1,'` ;; esac; \
 	for dir in tmp-foo intl $(SUBDIRS); do \
 	  if [ "`echo $$dir/*$(objext)`" != "$$dir/*$(objext)" ] ; then \
 	    for file in $$dir/*$(objext); do \
-	      tail +16c ./$$file > tmp-foo1; \
-	      tail +16c stage$$stage/$$file > tmp-foo2 \
+	      tail -c +16 ./$$file > tmp-foo1; \
+	      tail -c +16 stage$$stage/$$file > tmp-foo2 \
 	        && (cmp tmp-foo1 tmp-foo2 > /dev/null 2>&1 || echo $$file differs >> .bad_compare) || true; \
 	    done; \
 	  else true; fi; \
diff -urpN gcc-3.3.1/gcc/configure gcc-3.3.1.coreutils/gcc/configure
--- gcc-3.3.1/gcc/configure	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/gcc/configure	2003-09-06 22:56:10.245543480 +0200
@@ -7203,7 +7203,7 @@ elif test x$gcc_cv_as != x; then
  	# GNU LD versions before 2.12.1 have buggy support for STV_HIDDEN.
 	# This is irritatingly difficult to feature test for.  Look for 
 	# the date string after the version number.
-	ld_ver=`$gcc_cv_ld --version 2>/dev/null | head -1`
+	ld_ver=`$gcc_cv_ld --version 2>/dev/null | head -n 1`
 	if echo "$ld_ver" | grep GNU > /dev/null; then
                ld_vers=`echo $ld_ver | sed -n 's,^.*[  ]\([0-9][0-9]*\.[0-9][0-9]*\(\|\.[0-9][0-9]*\(\|\.[0-9][0-9]*\)\)\)\([  ].*\|\)$,\1,p'`
                ld_date=`echo $ld_ver | sed -n 's,^.*\([2-9][0-9][0-9][0-9]\)[-]*\([01][0-9]\)[-]*\([0-3][0-9]\).*$,\1\2\3,p'`
@@ -7279,7 +7279,7 @@ EOF
 		# ??? There exists an elf-specific test that will crash
 		# the assembler.  Perhaps it's better to figure out whether
 		# arbitrary sections are supported and try the test.
-		as_ver=`$gcc_cv_as --version 2>/dev/null | head -1`
+		as_ver=`$gcc_cv_as --version 2>/dev/null | head -n 1`
 		if echo "$as_ver" | grep GNU > /dev/null; then
 			as_ver=`echo $as_ver | sed -e 's/GNU assembler \([0-9.][0-9.]*\).*/\1/'`
 			as_major=`echo $as_ver | sed 's/\..*//'`
@@ -7308,7 +7308,7 @@ if test x$gcc_cv_gas_major_version != x 
   fi
 elif test x$gcc_cv_as != x; then
 	# Check if this is GAS.
-	as_ver=`$gcc_cv_as --version < /dev/null 2> /dev/null | head -1`
+	as_ver=`$gcc_cv_as --version < /dev/null 2> /dev/null | head -n 1`
 	rm -f a.out 2> /dev/null
 	if echo "$as_ver" | grep GNU > /dev/null; then
 		# Versions up to and including 2.11.0 may mis-optimize
@@ -7361,7 +7361,7 @@ EOF
 		# and we got the correct data, then succeed.
 		if $gcc_cv_as -o conftest.o conftest.s > /dev/null 2>&1 \
 		   && $gcc_cv_objdump -s -j .eh_frame conftest.o 2>/dev/null \
-		      | tail -3 > conftest.got \
+		      | tail -n 3 > conftest.got \
 		   && { cmp conftest.lit conftest.got > /dev/null 2>&1 \
 		        || cmp conftest.big conftest.got > /dev/null 2>&1; }
 		then
diff -urpN gcc-3.3.1/gcc/configure.in gcc-3.3.1.coreutils/gcc/configure.in
--- gcc-3.3.1/gcc/configure.in	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/gcc/configure.in	2003-09-06 22:56:10.297535576 +0200
@@ -1639,7 +1639,7 @@ elif test x$gcc_cv_as != x; then
  	# GNU LD versions before 2.12.1 have buggy support for STV_HIDDEN.
 	# This is irritatingly difficult to feature test for.  Look for 
 	# the date string after the version number.
-	ld_ver=`$gcc_cv_ld --version 2>/dev/null | head -1`
+	ld_ver=`$gcc_cv_ld --version 2>/dev/null | head -n 1`
 	if echo "$ld_ver" | grep GNU > /dev/null; then
 changequote(,)dnl
                ld_vers=`echo $ld_ver | sed -n 's,^.*[  ]\([0-9][0-9]*\.[0-9][0-9]*\(\|\.[0-9][0-9]*\(\|\.[0-9][0-9]*\)\)\)\([  ].*\|\)$,\1,p'`
@@ -1714,7 +1714,7 @@ EOF
 		# ??? There exists an elf-specific test that will crash
 		# the assembler.  Perhaps it's better to figure out whether
 		# arbitrary sections are supported and try the test.
-		as_ver=`$gcc_cv_as --version 2>/dev/null | head -1`
+		as_ver=`$gcc_cv_as --version 2>/dev/null | head -n 1`
 		if echo "$as_ver" | grep GNU > /dev/null; then
 changequote(,)dnl
 			as_ver=`echo $as_ver | sed -e 's/GNU assembler \([0-9.][0-9.]*\).*/\1/'`
@@ -1742,7 +1742,7 @@ if test x$gcc_cv_gas_major_version != x 
   fi
 elif test x$gcc_cv_as != x; then
 	# Check if this is GAS.
-	as_ver=`$gcc_cv_as --version < /dev/null 2> /dev/null | head -1`
+	as_ver=`$gcc_cv_as --version < /dev/null 2> /dev/null | head -n 1`
 	rm -f a.out 2> /dev/null
 	if echo "$as_ver" | grep GNU > /dev/null; then
 		# Versions up to and including 2.11.0 may mis-optimize
@@ -1795,7 +1795,7 @@ EOF
 		# and we got the correct data, then succeed.
 		if $gcc_cv_as -o conftest.o conftest.s > /dev/null 2>&1 \
 		   && $gcc_cv_objdump -s -j .eh_frame conftest.o 2>/dev/null \
-		      | tail -3 > conftest.got \
+		      | tail -n 3 > conftest.got \
 		   && { cmp conftest.lit conftest.got > /dev/null 2>&1 \
 		        || cmp conftest.big conftest.got > /dev/null 2>&1; }
 		then
diff -urpN gcc-3.3.1/libjava/acinclude.m4 gcc-3.3.1.coreutils/libjava/acinclude.m4
--- gcc-3.3.1/libjava/acinclude.m4	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/libjava/acinclude.m4	2003-09-06 22:55:58.872272480 +0200
@@ -279,8 +279,8 @@ AC_DEFUN([CHECK_FOR_BROKEN_MINGW_LD],
 AC_MSG_CHECKING(whether 'ld' is at least 2.13)
 LD_PROG=`$CC --print-prog-name=ld`
 LD_VERSION=`$LD_PROG --version`
-LD_VERSION_MAJOR=`echo "$LD_VERSION" | head -1 | cut -d '.' -f 1 | cut -d ' ' -f 4`
-LD_VERSION_MINOR=`echo "$LD_VERSION" | head -1 | cut -d '.' -f 2`
+LD_VERSION_MAJOR=`echo "$LD_VERSION" | head -n 1 | cut -d '.' -f 1 | cut -d ' ' -f 4`
+LD_VERSION_MINOR=`echo "$LD_VERSION" | head -n 1 | cut -d '.' -f 2`
 if expr "$LD_VERSION_MAJOR" \> 2 > /dev/null; then
   LD_OK="ok"
 else
diff -urpN gcc-3.3.1/libjava/aclocal.m4 gcc-3.3.1.coreutils/libjava/aclocal.m4
--- gcc-3.3.1/libjava/aclocal.m4	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/libjava/aclocal.m4	2003-09-06 22:55:58.881271112 +0200
@@ -291,8 +291,8 @@ AC_DEFUN([CHECK_FOR_BROKEN_MINGW_LD],
 AC_MSG_CHECKING(whether 'ld' is at least 2.13)
 LD_PROG=`$CC --print-prog-name=ld`
 LD_VERSION=`$LD_PROG --version`
-LD_VERSION_MAJOR=`echo "$LD_VERSION" | head -1 | cut -d '.' -f 1 | cut -d ' ' -f 4`
-LD_VERSION_MINOR=`echo "$LD_VERSION" | head -1 | cut -d '.' -f 2`
+LD_VERSION_MAJOR=`echo "$LD_VERSION" | head -n 1 | cut -d '.' -f 1 | cut -d ' ' -f 4`
+LD_VERSION_MINOR=`echo "$LD_VERSION" | head -n 1 | cut -d '.' -f 2`
 if expr "$LD_VERSION_MAJOR" \> 2 > /dev/null; then
   LD_OK="ok"
 else
diff -urpN gcc-3.3.1/libjava/configure gcc-3.3.1.coreutils/libjava/configure
--- gcc-3.3.1/libjava/configure	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/libjava/configure	2003-09-06 22:55:58.865273544 +0200
@@ -2965,8 +2965,8 @@ echo $ac_n "checking whether 'ld' is at 
 echo "configure:2966: checking whether 'ld' is at least 2.13" >&5
 LD_PROG=`$CC --print-prog-name=ld`
 LD_VERSION=`$LD_PROG --version`
-LD_VERSION_MAJOR=`echo "$LD_VERSION" | head -1 | cut -d '.' -f 1 | cut -d ' ' -f 4`
-LD_VERSION_MINOR=`echo "$LD_VERSION" | head -1 | cut -d '.' -f 2`
+LD_VERSION_MAJOR=`echo "$LD_VERSION" | head -n 1 | cut -d '.' -f 1 | cut -d ' ' -f 4`
+LD_VERSION_MINOR=`echo "$LD_VERSION" | head -n 1 | cut -d '.' -f 2`
 if expr "$LD_VERSION_MAJOR" \> 2 > /dev/null; then
   LD_OK="ok"
 else
diff -urpN gcc-3.3.1/libstdc++-v3/acinclude.m4 gcc-3.3.1.coreutils/libstdc++-v3/acinclude.m4
--- gcc-3.3.1/libstdc++-v3/acinclude.m4	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/libstdc++-v3/acinclude.m4	2003-09-06 22:55:59.369196936 +0200
@@ -315,7 +315,7 @@ AC_DEFUN(GLIBCPP_CHECK_LINKER_FEATURES, 
   # Start by getting the version number.  I think the libtool test already
   # does some of this, but throws away the result.
   changequote(,)
-  ldver=`$LD --version 2>/dev/null | head -1 | \
+  ldver=`$LD --version 2>/dev/null | head -n 1 | \
          sed -e 's/GNU ld version \([0-9.][0-9.]*\).*/\1/'`
   changequote([,])
   glibcpp_gnu_ld_version=`echo $ldver | \
diff -urpN gcc-3.3.1/libstdc++-v3/aclocal.m4 gcc-3.3.1.coreutils/libstdc++-v3/aclocal.m4
--- gcc-3.3.1/libstdc++-v3/aclocal.m4	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/libstdc++-v3/aclocal.m4	2003-09-06 22:55:59.425188424 +0200
@@ -327,7 +327,7 @@ AC_DEFUN(GLIBCPP_CHECK_LINKER_FEATURES, 
   # Start by getting the version number.  I think the libtool test already
   # does some of this, but throws away the result.
   changequote(,)
-  ldver=`$LD --version 2>/dev/null | head -1 | \
+  ldver=`$LD --version 2>/dev/null | head -n 1 | \
          sed -e 's/GNU ld version \([0-9.][0-9.]*\).*/\1/'`
   changequote([,])
   glibcpp_gnu_ld_version=`echo $ldver | \
diff -urpN gcc-3.3.1/libstdc++-v3/configure gcc-3.3.1.coreutils/libstdc++-v3/configure
--- gcc-3.3.1/libstdc++-v3/configure	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/libstdc++-v3/configure	2003-09-06 22:55:59.310205904 +0200
@@ -4285,7 +4285,7 @@ done
   # Start by getting the version number.  I think the libtool test already
   # does some of this, but throws away the result.
   
-  ldver=`$LD --version 2>/dev/null | head -1 | \
+  ldver=`$LD --version 2>/dev/null | head -n 1 | \
          sed -e 's/GNU ld version \([0-9.][0-9.]*\).*/\1/'`
   
   glibcpp_gnu_ld_version=`echo $ldver | \
@@ -5285,7 +5285,7 @@ done
   # Start by getting the version number.  I think the libtool test already
   # does some of this, but throws away the result.
   
-  ldver=`$LD --version 2>/dev/null | head -1 | \
+  ldver=`$LD --version 2>/dev/null | head -n 1 | \
          sed -e 's/GNU ld version \([0-9.][0-9.]*\).*/\1/'`
   
   glibcpp_gnu_ld_version=`echo $ldver | \
@@ -6265,7 +6265,7 @@ done
   # Start by getting the version number.  I think the libtool test already
   # does some of this, but throws away the result.
   
-  ldver=`$LD --version 2>/dev/null | head -1 | \
+  ldver=`$LD --version 2>/dev/null | head -n 1 | \
          sed -e 's/GNU ld version \([0-9.][0-9.]*\).*/\1/'`
   
   glibcpp_gnu_ld_version=`echo $ldver | \
@@ -7255,7 +7255,7 @@ done
   # Start by getting the version number.  I think the libtool test already
   # does some of this, but throws away the result.
   
-  ldver=`$LD --version 2>/dev/null | head -1 | \
+  ldver=`$LD --version 2>/dev/null | head -n 1 | \
          sed -e 's/GNU ld version \([0-9.][0-9.]*\).*/\1/'`
   
   glibcpp_gnu_ld_version=`echo $ldver | \
@@ -8259,7 +8259,7 @@ done
   # Start by getting the version number.  I think the libtool test already
   # does some of this, but throws away the result.
   
-  ldver=`$LD --version 2>/dev/null | head -1 | \
+  ldver=`$LD --version 2>/dev/null | head -n 1 | \
          sed -e 's/GNU ld version \([0-9.][0-9.]*\).*/\1/'`
   
   glibcpp_gnu_ld_version=`echo $ldver | \
@@ -9154,7 +9154,7 @@ fi	
   # Start by getting the version number.  I think the libtool test already
   # does some of this, but throws away the result.
   
-  ldver=`$LD --version 2>/dev/null | head -1 | \
+  ldver=`$LD --version 2>/dev/null | head -n 1 | \
          sed -e 's/GNU ld version \([0-9.][0-9.]*\).*/\1/'`
   
   glibcpp_gnu_ld_version=`echo $ldver | \
@@ -10638,7 +10638,7 @@ cross_compiling=$ac_cv_prog_cc_cross
   # Start by getting the version number.  I think the libtool test already
   # does some of this, but throws away the result.
   
-  ldver=`$LD --version 2>/dev/null | head -1 | \
+  ldver=`$LD --version 2>/dev/null | head -n 1 | \
          sed -e 's/GNU ld version \([0-9.][0-9.]*\).*/\1/'`
   
   glibcpp_gnu_ld_version=`echo $ldver | \
diff -urpN gcc-3.3.1/libstdc++-v3/mkcheck.in gcc-3.3.1.coreutils/libstdc++-v3/mkcheck.in
--- gcc-3.3.1/libstdc++-v3/mkcheck.in	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/libstdc++-v3/mkcheck.in	2003-09-06 22:56:10.307534056 +0200
@@ -181,9 +181,9 @@ setup_size_command()
     function size_command()
     {
       case $1 in
-        TEXT)  TEXT=$(size -B $EXENAME | tail -1 | awk '{print $1}')  ;;
-        DATA)  DATA=$(size -B $EXENAME | tail -1 | awk '{print $2}')  ;;
-        SIZE)  SIZE=$(size -B $EXENAME | tail -1 | awk '{print $4}')  ;;
+        TEXT)  TEXT=$(size -B $EXENAME | tail -n 1 | awk '{print $1}')  ;;
+        DATA)  DATA=$(size -B $EXENAME | tail -n 1 | awk '{print $2}')  ;;
+        SIZE)  SIZE=$(size -B $EXENAME | tail -n 1 | awk '{print $4}')  ;;
       esac
     }
   else
diff -urpN gcc-3.3.1/ltcf-c.sh gcc-3.3.1.coreutils/ltcf-c.sh
--- gcc-3.3.1/ltcf-c.sh	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/ltcf-c.sh	2003-09-06 22:55:59.443185688 +0200
@@ -153,7 +153,7 @@ EOF
     # If the export-symbols file already is a .def file (1st line
     # is EXPORTS), use it as is.
     # If DATA tags from a recent dlltool are present, honour them!
-    archive_expsym_cmds='if test "x`head -1 $export_symbols`" = xEXPORTS; then
+    archive_expsym_cmds='if test "x`head -n 1 $export_symbols`" = xEXPORTS; then
         cp $export_symbols $output_objdir/$soname-def;
       else
         echo EXPORTS > $output_objdir/$soname-def;
diff -urpN gcc-3.3.1/ltcf-gcj.sh gcc-3.3.1.coreutils/ltcf-gcj.sh
--- gcc-3.3.1/ltcf-gcj.sh	2003-09-06 22:47:26.000000000 +0200
+++ gcc-3.3.1.coreutils/ltcf-gcj.sh	2003-09-06 22:55:59.458183408 +0200
@@ -156,7 +156,7 @@ EOF
     # If the export-symbols file already is a .def file (1st line
     # is EXPORTS), use it as is.
     # If DATA tags from a recent dlltool are present, honour them!
-    archive_expsym_cmds='if test "x`head -1 $export_symbols`" = xEXPORTS; then
+    archive_expsym_cmds='if test "x`head -n 1 $export_symbols`" = xEXPORTS; then
         cp $export_symbols $output_objdir/$soname-def;
       else
         echo EXPORTS > $output_objdir/$soname-def;
