From ddf4aa12e75d56f33bc225d55c42749e92c96a24 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 20 Aug 2017 15:39:25 -0400
Subject: [PATCH 011/108] enable SLAB_FREELIST_HARDENED by default

---
 init/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/Kconfig b/init/Kconfig
index 32e37bdb8019..10f6ab320171 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1925,6 +1925,7 @@ config SLAB_FREELIST_RANDOM
 config SLAB_FREELIST_HARDENED
 	bool "Harden slab freelist metadata"
 	depends on SLAB || SLUB
+	default y
 	help
 	  Many kernel heap attacks try to target slab cache metadata and
 	  other infrastructure. This options makes minor performance
-- 
2.29.2

