From b5738a36f247c9371dd5271e2d1997d313ef9e90 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 7 May 2017 18:28:33 -0400
Subject: [PATCH 031/108] disable COMPAT_BRK by default

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 37c7d69a8071..5885c02eacfa 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1853,7 +1853,6 @@ config SLUB_MEMCG_SYSFS_ON
 
 config COMPAT_BRK
 	bool "Disable heap randomization"
-	default y
 	help
 	  Randomizing heap placement makes heap exploits harder, but it
 	  also breaks ancient binaries (including anything libc5 based).
-- 
2.29.2

