From 56e9a492bd3221cc5d27452c9d6825c9d114cca2 Mon Sep 17 00:00:00 2001
From: Dave Hughes <davidhughes205@gmail.com>
Date: Tue, 28 Sep 2021 21:08:18 +0100
Subject: [PATCH] python/ego/profile.py: fix per arch mix-ins

---
 python/ego/profile.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/python/ego/profile.py b/python/ego/profile.py
index 8d560b1..6ec44e7 100644
--- a/python/ego/profile.py
+++ b/python/ego/profile.py
@@ -254,7 +254,7 @@ class ProfileCatalog:
 			if key == ProfileType.SUBARCH:
 				dirlist = [self.json_info[str(ProfileType.ARCH)] + "/" + arch + "/subarch"]
 			elif key == ProfileType.MIX_IN:
-				dirlist = [self.json_info[str(ProfileType.MIX_IN)] + "/" + arch + "/mix-ins"]
+				dirlist = [self.json_info[str(ProfileType.ARCH)] + "/" + arch + "/mix-ins"]
 
 		if str(key) in self.json_info:
 			dirlist += [self.json_info[str(key)]]
-- 
2.33.0

