From 301f9a33a59067ca199d9d071d45bc67935daedf Mon Sep 17 00:00:00 2001
From: Dave Hughes <davidhughes205@gmail.com>
Date: Tue, 28 Sep 2021 21:31:01 +0100
Subject: [PATCH] modules/sync.ego: fix for update_repos_conf

---
 modules/sync.ego | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/modules/sync.ego b/modules/sync.ego
index 789c800..0a52867 100644
--- a/modules/sync.ego
+++ b/modules/sync.ego
@@ -47,7 +47,7 @@ class Module(EgoModule):
 				os.makedirs(os.path.dirname(root))
 			self._kits_root = root
 		return self._kits_root
-	
+
 	def _do_package_moves(self):
 		import portage
 		from portage._global_updates import _global_updates
@@ -58,7 +58,7 @@ class Module(EgoModule):
 			emerge_config.target_config.mtimedb.commit()
 		except portage.exception.ParseError as e:
 			Output.warning("Unable to perform package moves due to the following parse error: " + str(e))
-			
+
 	def add_arguments(self, parser):
 		parser.add_argument('--kits', action='store_true', default=True, help="Sync kits.")
 		parser.add_argument('--no-kits', action='store_false', dest='kits', help="Disable kits syncing.")
@@ -186,6 +186,7 @@ class Module(EgoModule):
 				if kit_name == "core-kit":
 					f.write("""[DEFAULT]
 main-repo = core-kit
+force = aliases
 
 """)
 				f.write("""[%s]
-- 
2.33.0

