From 3c1264bd1cd0538dbe75dc894197ecaf4886cd21 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sat, 27 May 2017 07:28:10 -0400
Subject: [PATCH 025/104] disable DEVPORT by default

---
 drivers/char/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/char/Kconfig b/drivers/char/Kconfig
index df0fc997dc3e..0c02ae88f792 100644
--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -514,7 +514,6 @@ config TELCLOCK
 config DEVPORT
 	bool "/dev/port character device"
 	depends on ISA || PCI
-	default y
 	help
 	  Say Y here if you want to support the /dev/port device. The /dev/port
 	  device is similar to /dev/mem, but for I/O ports.
-- 
2.28.0

