From ad8cf81234f6aa40d61aac4d6a6d8da50be2fcbf Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 7 May 2017 18:28:33 -0400
Subject: [PATCH 032/104] disable COMPAT_BRK by default

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 3e04647c3932..17a5382fdd93 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1748,7 +1748,6 @@ config SLUB_MEMCG_SYSFS_ON
 
 config COMPAT_BRK
 	bool "Disable heap randomization"
-	default y
 	help
 	  Randomizing heap placement makes heap exploits harder, but it
 	  also breaks ancient binaries (including anything libc5 based).
-- 
2.28.0

