From 20d31776fce9fe3c4381eb4c943919cc4a16b51d Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Fri, 5 Jan 2018 12:43:49 -0500
Subject: [PATCH 030/103] enable IO_STRICT_DEVMEM by default

---
 lib/Kconfig.debug | 1 +
 1 file changed, 1 insertion(+)

diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index 4f9ce2bd9e61..ceed9e31fb19 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -1543,6 +1543,7 @@ config STRICT_DEVMEM
 config IO_STRICT_DEVMEM
 	bool "Filter I/O access to /dev/mem"
 	depends on STRICT_DEVMEM
+	default y
 	help
 	  If this option is disabled, you allow userspace (root) access to all
 	  io-memory regardless of whether a driver is actively using that
-- 
2.28.0

