From c3899f4ac55a90679c7b1e1175edd30a3dbf077c Mon Sep 17 00:00:00 2001
From: Bernhard40 <32568352+Bernhard40@users.noreply.github.com>
Date: Fri, 6 Oct 2017 10:24:10 +0000
Subject: [PATCH 022/113] disable AIO by default

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 8605f3e78e47..21f0b6926cf3 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1592,7 +1592,6 @@ config SHMEM
 
 config AIO
 	bool "Enable AIO support"
-	default y
 	help
 	  This option enables POSIX asynchronous I/O which may by used
 	  by some high performance threaded applications. Disabling
-- 
2.30.1

