--- branches/gcc-4_1-branch/gcc/config/pa/pa.c	2006/11/19 19:06:09	118995
+++ branches/gcc-4_1-branch/gcc/config/pa/pa.c	2006/11/24 04:35:06	119147
@@ -4323,8 +4323,10 @@
 		 GEN_INT (0x00011820), NE, NULL_RTX, SImode, 1);
   emit_jump_insn (gen_bne (label));
 
+  /* 0xe0400002 must be specified as -532676606 so that it won't be
+     rejected as an invalid immediate operand on 64-bit hosts.  */
   emit_cmp_insn (gen_rtx_MEM (SImode, plus_constant (ins, 12)),
-		 GEN_INT (0xe0400002), NE, NULL_RTX, SImode, 1);
+		 GEN_INT (-532676606), NE, NULL_RTX, SImode, 1);
 
   /* If there is no export stub then just use the value saved from
      the return pointer register.  */
