http://gcc.gnu.org/PR27830

--- branches/gcc-4_1-branch/gcc/tree-inline.c	2006/06/16 10:35:28	114715
+++ branches/gcc-4_1-branch/gcc/tree-inline.c	2006/06/16 12:03:24	114716
@@ -709,7 +709,12 @@
       else if (TREE_CODE (*tp) == ADDR_EXPR)
 	{
 	  walk_tree (&TREE_OPERAND (*tp, 0), copy_body_r, id, NULL);
-	  recompute_tree_invarant_for_addr_expr (*tp);
+	  /* Handle the case where we substituted an INDIRECT_REF
+	     into the operand of the ADDR_EXPR.  */
+	  if (TREE_CODE (TREE_OPERAND (*tp, 0)) == INDIRECT_REF)
+	    *tp = TREE_OPERAND (TREE_OPERAND (*tp, 0), 0);
+	  else
+	    recompute_tree_invarant_for_addr_expr (*tp);
 	  *walk_subtrees = 0;
 	}
     }
