--- gcc-3.3.5/gcc/config/pa/pa-linux.h.upstream~	Thu Nov 18 23:11:39 2004
+++ gcc-3.3.5/gcc/config/pa/pa-linux.h	Thu Nov 18 23:17:02 2004
@@ -60,6 +60,11 @@
 	builtin_define ("__gnu_linux__");	\
 	builtin_define_std ("linux");		\
 	builtin_define_std ("unix");		\
+	if (flag_pic)				\
+	  {					\
+	    builtin_define ("__PIC__");		\
+	    builtin_define ("__pic__");		\
+	  }					\
 	builtin_assert ("machine=bigendian");	\
 	builtin_assert ("system=posix");	\
 	builtin_assert ("system=unix");		\
@@ -67,7 +72,7 @@
   while (0)
 
 #undef CPP_SPEC
-#define CPP_SPEC "%{fPIC|fpic|fPIE|fpie:-D__PIC__ -D__pic__} %{posix:-D_POSIX_SOURCE}"
+#define CPP_SPEC "%{posix:-D_POSIX_SOURCE}"
 
 #undef	LIB_SPEC
 #define LIB_SPEC \
--- gcc-3.3.5/gcc/config/arm/linux-gas.h.upstream~	Thu Nov 18 23:11:39 2004
+++ gcc-3.3.5/gcc/config/arm/linux-gas.h	Thu Nov 18 23:11:39 2004
@@ -31,7 +31,7 @@
 #define DEFAULT_SIGNED_CHAR 0
 
 #undef  SUBTARGET_CPP_SPEC
-#define SUBTARGET_CPP_SPEC  "%{posix:-D_POSIX_SOURCE} %{fPIC|fPIE:-D__PIC__ -D__pic__} %{fpic|fpie:-D__PIC__ -D__pic__}"
+#define SUBTARGET_CPP_SPEC  "%{posix:-D_POSIX_SOURCE}"
 
 #undef  SIZE_TYPE
 #define SIZE_TYPE "unsigned int"
--- gcc-3.3.5/gcc/config/arm/linux-elf.h.upstream~	Thu Nov 18 23:11:39 2004
+++ gcc-3.3.5/gcc/config/arm/linux-elf.h	Thu Nov 18 23:17:20 2004
@@ -110,6 +110,11 @@
 	builtin_define_std ("linux");		\
 	builtin_define ("__gnu_linux__");	\
 	builtin_define ("__ELF__");		\
+	if (flag_pic)				\
+	  {					\
+	    builtin_define ("__PIC__");		\
+	    builtin_define ("__pic__");		\
+	  }					\
 	builtin_assert ("system=unix");		\
 	builtin_assert ("system=posix");	\
     } while (0)
--- gcc-3.3.5/gcc/config/alpha/linux.h.upstream~	Wed Oct  1 18:58:40 2003
+++ gcc-3.3.5/gcc/config/alpha/linux.h	Thu Nov 18 23:11:39 2004
@@ -33,6 +33,11 @@
 	/* The GNU C++ standard library requires this.  */	\
 	if (c_language == clk_cplusplus)			\
 	  builtin_define ("_GNU_SOURCE");			\
+	if (flag_pic)						\
+	  {							\
+		builtin_define ("__PIC__");			\
+		builtin_define ("__pic__");			\
+	  }							\
     } while (0)
 
 #undef LIB_SPEC
