--- gcc-3.3.5/gcc/config/rs6000/sysv4.h.upstream~	Fri Nov 19 00:28:57 2004
+++ gcc-3.3.5/gcc/config/rs6000/sysv4.h	Fri Nov 19 10:41:44 2004
@@ -810,6 +810,22 @@
 #define	TARGET_VERSION fprintf (stderr, " (PowerPC System V.4)");
 #endif
 
+#define TARGET_OS_SYSV_CPP_BUILTINS()		\
+  do						\
+    {						\
+      if (flag_pic == 1)			\
+	{					\
+	  builtin_define ("__pic__=1");		\
+	  builtin_define ("__PIC__=1");		\
+	}					\
+      else if (flag_pic == 2)			\
+	{					\
+	  builtin_define ("__pic__=2");		\
+	  builtin_define ("__PIC__=2");		\
+	}					\
+    }						\
+  while (0)
+
 #ifndef	TARGET_OS_CPP_BUILTINS
 #define TARGET_OS_CPP_BUILTINS()          \
   do                                      \
@@ -821,6 +837,7 @@
       builtin_assert ("system=svr4");     \
       builtin_assert ("cpu=powerpc");     \
       builtin_assert ("machine=powerpc"); \
+      TARGET_OS_SYSV_CPP_BUILTINS ();     \
     }                                     \
   while (0)
 #endif
@@ -976,10 +993,7 @@
 
 #define LINK_OS_DEFAULT_SPEC ""
 
-#define CPP_SYSV_SPEC \
-"%{mrelocatable*: -D_RELOCATABLE} \
-%{fpic: -D__PIC__=1 -D__pic__=1} \
-%{!fpic: %{fPIC: -D__PIC__=2 -D__pic__=2}}"
+#define CPP_SYSV_SPEC "%{mrelocatable*: -D_RELOCATABLE}"
 
 /* Override rs6000.h definition.  */
 #undef	CPP_SPEC
