--- gcc-3.3.3/gcc/config/sparc/linux.h.sparcpie	Tue May  4 15:29:07 2004
+++ gcc-3.3.3/gcc/config/sparc/linux.h	Tue May  4 17:34:52 2004
@@ -47,18 +47,19 @@
      %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:crt1.o%s}}}\
    crti.o%s %{!shared:crtbegin.o%s} %{shared:crtbeginS.o%s}"
 #else
-#if defined HAVE_LD_PIE
-#define STARTFILE_SPEC \
-  "%{!shared: \
-     %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{pie:Scrt1.o%s}%{!pie:crt1.o%s}}}}\
-   crti.o%s %{static:crtbeginT.o%s}\
-   %{!static:%{!shared:%{!pie:crtbegin.o%s}} %{shared|pie:crtbeginS.o%s}}"
-#else
-#define STARTFILE_SPEC \
-  "%{!shared: \
-     %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:crt1.o%s}}}\
-   crti.o%s %{static:crtbeginT.o%s}\
-   %{!static:%{!shared:%{!pie:crtbegin.o%s}} %{shared|pie:crtbeginS.o%s}}"
+#ifdef HAVE_LD_PIE
+#define STARTFILE_SPEC "\
+%{!shared: \
+%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{static:crt1.o%s} \
+%{!static:%{pie:Scrt1.o%s}%{!pie:crt1.o%s}} }}} \
+crti.o%s %{shared:crtbeginS.o%s} %{!shared:%{static:crtbeginT.o%s} \
+%{!static:%{pie:crtbeginS.o%s}%{!pie:crtbegin.o%s}} }"
+#else
+#define STARTFILE_SPEC "\
+%{!shared: \
+%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:crt1.o%s}}} \
+crti.o%s %{shared:crtbeginS.o%s} %{!shared:%{static:crtbeginT.o%s} \
+%{!static:crtbegin.o%s} }"
 #endif
 #endif
 
@@ -69,9 +70,16 @@
    GNU/Linux "finalizer" file, `crtn.o'.  */
 
 #undef  ENDFILE_SPEC
-#define ENDFILE_SPEC \
-  "%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
-   %{!shared:%{!pie:crtend.o%s}} %{shared|pie:crtendS.o%s} crtn.o%s"
+#ifdef HAVE_LD_PIE
+#define ENDFILE_SPEC "\
+%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
+%{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
+%{!static:%{pie:crtendS.o%s}%{!pie:crtend.o%s}} } crtn.o%s"
+#else
+#define ENDFILE_SPEC "\
+%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
+%{shared:crtendS.o%s} %{!shared:crtend.o%s} crtn.o%s"
+#endif
 
 /* This is for -profile to use -lc_p instead of -lc.  */
 #undef	CC1_SPEC
