--- gcc-3.3.3/gcc/config/sparc/linux64.h.sparc64pie	Tue May  4 15:29:07 2004
+++ gcc-3.3.3/gcc/config/sparc/linux64.h	Tue May  4 17:29:02 2004
@@ -56,15 +56,16 @@
 #undef  STARTFILE_SPEC
 
 #ifdef HAVE_LD_PIE
-#define STARTFILE_SPEC \
-  "%{!shared:%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{pie:Scrt1.o%s}%{!pie:crt1.o%s}}}}\
-   crti.o%s %{static:crtbeginT.o%s}\
-   %{!static:%{!shared:%{!pie:crtbegin.o%s}} %{shared|pie:crtbeginS.o%s}}"
+#define STARTFILE_SPEC "\
+%{!shared:%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{static:crt1.o%s} \
+%{!static:%{pie:Scrt1.o%s}%{!pie:crt1.o%s}} }}} \
+crti.o%s %{shared:crtbeginS.o%s} %{!shared:%{static:crtbeginT.o%s} \
+%{!static:%{pie:crtbeginS.o%s}%{!pie:crtbegin.o%s}} }"
 #else
-#define STARTFILE_SPEC \
-  "%{!shared:%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:crt1.o%s}}}\
-   crti.o%s %{static:crtbeginT.o%s}\
-   %{!static:%{!shared:%{!pie:crtbegin.o%s}} %{shared|pie:crtbeginS.o%s}}"
+#define STARTFILE_SPEC "\
+%{!shared:%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:crt1.o%s}}} \
+crti.o%s %{shared:crtbeginS.o%s} %{!shared:%{static:crtbeginT.o%s} \
+%{!static:crtbegin.o%s} }"
 #endif
 
 /* Provide a ENDFILE_SPEC appropriate for GNU/Linux.  Here we tack on
@@ -75,9 +76,16 @@
 
 #undef  ENDFILE_SPEC
 
-#define ENDFILE_SPEC \
-  "%{!shared:%{!pie:crtend.o%s}} %{shared|pie:crtendS.o%s} crtn.o%s\
-   %{ffast-math|funsafe-math-optimizations:crtfastmath.o%s}"
+#ifdef HAVE_LD_PIE
+#define ENDFILE_SPEC "\
+%{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
+%{!static:%{pie:crtendS.o%s}%{!pie:crtend.o%s}} } crtn.o%s \
+%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s}"
+#else
+#define ENDFILE_SPEC "\
+%{shared:crtendS.o%s} %{!shared:crtend.o%s} crtn.o%s \
+%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s}"
+#endif
 
 /* The GNU C++ standard library requires that these macros be defined.  */
 #undef CPLUSPLUS_CPP_SPEC
