--- gcc-3.3.3/gcc/config/rs6000/sysv4.h.piecor	Mon May  3 16:15:37 2004
+++ gcc-3.3.3/gcc/config/rs6000/sysv4.h	Mon May  3 16:15:39 2004
@@ -1140,23 +1140,32 @@
 %{mnewlib: ecrti.o%s} %{!mnewlib: crti.o%s} \
 %{!shared:crtbegin.o%s} %{shared:crtbeginS.o%s}"
 #else
-#if defined HAVE_LD_PIE
+#ifdef HAVE_LD_PIE
 #define	STARTFILE_LINUX_SPEC "\
-%{!shared: %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{pie:Scrt1.o%s}%{!pie:crt1.o%s}}}} \
+%{!shared: %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{static:crt1.o%s} \
+%{!static:%{pie:Scrt1.o%s}%{!pie:crt1.o%s}} }}} \
 %{mnewlib: ecrti.o%s} %{!mnewlib: crti.o%s} \
-%{static:crtbeginT.o%s} \
-%{!static:%{!shared:%{!pie:crtbegin.o%s}} %{shared|pie:crtbeginS.o%s}}"
+%{shared:crtbeginS.o%s} %{!shared:%{static:crtbeginT.o%s} \
+%{!static:%{pie:crtbeginS.o%s}%{!pie:crtbegin.o%s}} }"
 #else
 #define	STARTFILE_LINUX_SPEC "\
 %{!shared: %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:crt1.o%s}}} \
 %{mnewlib: ecrti.o%s} %{!mnewlib: crti.o%s} \
-%{static:crtbeginT.o%s} \
-%{!static:%{!shared:%{!pie:crtbegin.o%s}} %{shared|pie:crtbeginS.o%s}}"
+%{shared:crtbeginS.o%s} %{!shared:%{static:crtbeginT.o%s} \
+%{!static:crtbegin.o%s} }"
 #endif
 #endif
 
-#define	ENDFILE_LINUX_SPEC "%{!shared:%{!pie:crtend.o%s}} %{shared|pie:crtendS.o%s} \
+#ifdef HAVE_LD_PIE
+#define	ENDFILE_LINUX_SPEC "\
+%{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
+%{!static:%{pie:crtendS.o%s}%{!pie:crtend.o%s}} } \
 %{mnewlib: ecrtn.o%s} %{!mnewlib: crtn.o%s}"
+#else
+#define	ENDFILE_LINUX_SPEC "\
+%{shared:crtendS.o%s} %{!shared:crtend.o%s} \
+%{mnewlib: ecrtn.o%s} %{!mnewlib: crtn.o%s}"
+#endif
 
 #define LINK_START_LINUX_SPEC ""
 
--- gcc-3.3.5/gcc/config/rs6000/linux64.h.piecor	Mon May  3 16:15:37 2004
+++ gcc-3.3.5/gcc/config/rs6000/linux64.h	Mon May  3 16:15:39 2004
@@ -165,14 +165,28 @@
 #endif
 
 #undef  STARTFILE_LINUX_SPEC
+#ifdef HAVE_LD_PIE
 #define STARTFILE_LINUX_SPEC "\
-%{!shared: %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:crt1.o%s}}} crti.o%s \
-%{static:crtbeginT.o%s} \
-%{!static:%{!shared:crtbegin.o%s} %{shared:crtbeginS.o%s}}"
+%{!shared: %{pg|p:gcrt1.o%s} %{!pg:%{!p:%{static:crt1.o%s} \
+%{!static:%{pie:Scrt1.o%s}%{!pie:crt1.o%s}} }}} crti.o%s \
+%{shared:crtbeginS.o%s} %{!shared:%{static:crtbeginT.o%s} \
+%{!static:%{pie:crtbeginS.o%s}%{!pie:crtbegin.o%s}}}"
+#else
+#define STARTFILE_LINUX_SPEC "\
+%{!shared: %{pg|p:gcrt1.o%s} %{!pg:%{!p:crt1.o%s}}} crti.o%s \
+%{shared:crtbeginS.o%s} %{!shared:%{static:crtbeginT.o%s} \
+%{!static:crtbegin.o%s} }"
+#endif
 
 #undef  ENDFILE_LINUX_SPEC
+#ifdef HAVE_LD_PIE
 #define ENDFILE_LINUX_SPEC "\
-%{!shared:crtend.o%s} %{shared:crtendS.o%s} crtn.o%s"
+%{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
+%{!static:%{pie:crtendS.o%s}%{!pie:crtend.o%s}}} crtn.o%s"
+#else
+#define ENDFILE_LINUX_SPEC "\
+%{shared:crtendS.o%s} %{!shared:crtend.o%s} crtn.o%s"
+#endif
 
 #undef  TOC_SECTION_ASM_OP
 #define TOC_SECTION_ASM_OP "\t.section\t\".toc\",\"aw\""
