--- gcc-3.3.3/gcc/config/arm/linux-elf.h.armpie	Tue May  4 16:06:13 2004
+++ gcc-3.3.3/gcc/config/arm/linux-elf.h	Tue May  4 16:30:09 2004
@@ -62,12 +62,21 @@
    object constructed before entering `main'. */
    
 #undef  STARTFILE_SPEC
-#define STARTFILE_SPEC \
-  "%{!shared: \
-     %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} \
-		       %{!p:%{profile:gcrt1.o%s} \
-			 %{!profile:crt1.o%s}}}} \
-   crti.o%s %{!shared:crtbegin.o%s} %{shared:crtbeginS.o%s}"
+#ifdef HAVE_LD_PIE
+#define STARTFILE_SPEC "\
+%{!shared: \
+%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} \
+%{!p:%{profile:gcrt1.o%s} %{!profile:%{static:crt1.o%s} \
+%{!static:%{pie:Scrt1.o%s}%{!pie:crt1.o%s}} }}}} \
+crti.o%s %{shared:crtbeginS.o%s} %{!shared:%{static:crtbegin.o%s} \
+%{!static:%{pie:crtbeginS.o%s}%{!pie:crtbegin.o%s}} }"
+#else
+#define STARTFILE_SPEC "\
+%{!shared: \
+%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} \
+%{!p:%{profile:gcrt1.o%s} %{!profile:crt1.o%s} }}} \
+crti.o%s %{shared:crtbeginS.o%s} %{!shared:crtbegin.o%s}"
+#endif
 
 /* Provide a ENDFILE_SPEC appropriate for GNU/Linux.  Here we tack on
    the GNU/Linux magical crtend.o file (see crtstuff.c) which
@@ -76,8 +85,14 @@
    GNU/Linux "finalizer" file, `crtn.o'.  */
 
 #undef  ENDFILE_SPEC
-#define ENDFILE_SPEC \
-  "%{!shared:crtend.o%s} %{shared:crtendS.o%s} crtn.o%s"
+#ifdef HAVE_LD_PIE
+#define ENDFILE_SPEC "\
+%{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
+%{!static:%{pie:crtendS.o%s}%{!pie:crtend.o%s}} } crtn.o%s"
+#else
+#define ENDFILE_SPEC "\
+%{shared:crtendS.o%s} %{!shared:crtend.o%s} crtn.o%s"
+#endif
 
 #undef  LINK_SPEC
 #ifdef USE_UCLIBC
