diff --git a/gcc/defaults.h b/gcc/defaults.h
index f1a38626624..158af7366bc 100644
--- a/gcc/defaults.h
+++ b/gcc/defaults.h
@@ -1365,8 +1365,12 @@ see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
 /* Default value for flag_stack_protect when flag_stack_protect is initialized to -1:
    --enable-default-ssp: Default flag_stack_protect to -fstack-protector-strong.
    --disable-default-ssp: Default flag_stack_protect to 0.
+   Hardened: Default flag_stack_protect to -fstack-protector-all.
  */
 #ifdef ENABLE_DEFAULT_SSP
+# ifdef ENABLE_DEFAULT_SSP_ALL
+#  define DEFAULT_FLAG_SSP 2
+# endif
 # ifndef DEFAULT_FLAG_SSP
 #  define DEFAULT_FLAG_SSP 3
 # endif
