--- branches/gcc-4_3-branch/gcc/emit-rtl.c	2008/06/23 13:06:15	137038
+++ branches/gcc-4_3-branch/gcc/emit-rtl.c	2008/06/23 13:16:07	137039
@@ -960,6 +960,12 @@
 {
   int offset;
 
+  /* Hard registers can be reused for multiple purposes within the same
+     function, so setting REG_ATTRS, REG_POINTER and REG_POINTER_ALIGN
+     on them is wrong.  */
+  if (HARD_REGISTER_P (reg))
+    return;
+
   offset = byte_lowpart_offset (GET_MODE (reg), GET_MODE (x));
   if (MEM_P (x))
     {
