From 8361ddec8f7997fb5c95d40f0d1738ae62c31929 Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Tue, 22 Dec 2020 23:40:09 +0100
Subject: [PATCH 045/110] stop hiding UID16 behind EXPERT

---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index fadc92229bf7..e863f3f4c39f 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1408,7 +1408,7 @@ menuconfig EXPERT
 	  Only use this if you really know what you are doing.
 
 config UID16
-	bool "Enable 16-bit UID system calls" if EXPERT
+	bool "Enable 16-bit UID system calls"
 	depends on HAVE_UID16 && MULTIUSER
 	default y
 	help
-- 
2.31.1

