From c36656d005e594f20261e9a2713f4d0ec6949791 Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Tue, 22 Dec 2020 23:36:54 +0100
Subject: [PATCH 044/112] disable SYSFS_SYSCALL by default

---
 init/Kconfig | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index 1ec6e3032160..d32804be1e31 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1434,13 +1434,12 @@ config SGETMASK_SYSCALL
 
 config SYSFS_SYSCALL
 	bool "Sysfs syscall support"
-	default y
 	help
 	  sys_sysfs is an obsolete system call no longer supported in libc.
 	  Note that disabling this option is more secure but might break
 	  compatibility with some systems.
 
-	  If unsure say Y here.
+	  If unsure say N here.
 
 config FHANDLE
 	bool "open by fhandle syscalls" if EXPERT
-- 
2.31.1

