--- gcc-3.3.3/gcc/config/arm/linux-elf.h.defpie	Tue May  4 20:38:17 2004
+++ gcc-3.3.3/gcc/config/arm/linux-elf.h	Tue May  4 20:45:37 2004
@@ -61,8 +61,16 @@
    provides part of the support for getting C++ file-scope static
    object constructed before entering `main'. */
    
+#undef  HARDENED_STARTFILE_SPEC
 #undef  STARTFILE_SPEC
 #ifdef HAVE_LD_PIE
+#define HARDENED_STARTFILE_SPEC "\
+%{!shared: \
+%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} \
+%{!p:%{profile:gcrt1.o%s} %{!profile:%{static:crt1.o%s} \
+%{!static:%{!nopie:Scrt1.o%s}%{nopie:crt1.o%s}} }}}} \
+crti.o%s %{shared:crtbeginS.o%s} %{!shared:%{static:crtbegin.o%s} \
+%{!static:%{!nopie:crtbeginS.o%s}%{nopie:crtbegin.o%s}} }"
 #define STARTFILE_SPEC "\
 %{!shared: \
 %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} \
@@ -84,8 +92,12 @@
    object constructed before entering `main', followed by a normal
    GNU/Linux "finalizer" file, `crtn.o'.  */
 
+#undef  HARDENED_ENDFILE_SPEC
 #undef  ENDFILE_SPEC
 #ifdef HAVE_LD_PIE
+#define HARDENED_ENDFILE_SPEC "\
+%{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
+%{!static:%{!nopie:crtendS.o%s}%{nopie:crtend.o%s}} } crtn.o%s"
 #define ENDFILE_SPEC "\
 %{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
 %{!static:%{pie:crtendS.o%s}%{!pie:crtend.o%s}} } crtn.o%s"
