--- gcc-3.3.5/gcc/config/rs6000/sysv4.h.def~	Tue Nov 30 16:02:43 2004
+++ gcc-3.3.5/gcc/config/rs6000/sysv4.h	Tue Nov 30 16:31:21 2004
@@ -859,6 +859,27 @@
     %{mcall-netbsd: -mbig} \
 }}}}"
 
+#ifdef HAVE_LD_PIE
+#if defined(EFAULT_PIE) || defined(EFAULT_PIE_SSP)
+#define	HARDENED_ASM_SPEC "%(asm_cpu) \
+%{.s: %{mregnames} %{mno-regnames}} %{.S: %{mregnames} %{mno-regnames}} \
+%{v:-V} %{Qy:} %{!Qn:-Qy} %{n} %{T} %{Ym,*} %{Yd,*} %{Wa,*:%*} \
+%{mrelocatable} %{mrelocatable-lib} \
+%{!D__KERNEL__: %{!fno-PIC:%{!fno-pic:%{!fno-PIE:%{!fno-pie: %{!nopie:-K PIC} }}}} } \
+%{memb} %{!memb: %{msdata: -memb} %{msdata=eabi: -memb}} \
+%{mlittle} %{mlittle-endian} %{mbig} %{mbig-endian} \
+%{!mlittle: %{!mlittle-endian: %{!mbig: %{!mbig-endian: \
+    %{mcall-freebsd: -mbig} \
+    %{mcall-i960-old: -mlittle} \
+    %{mcall-linux: -mbig} \
+    %{mcall-gnu: -mbig} \
+    %{mcall-netbsd: -mbig} \
+}}}}"
+#undef	ASM_SPEC
+#define	ASM_SPEC HARDENED_ASM_SPEC
+#endif
+#endif
+
 #define	CC1_ENDIAN_BIG_SPEC ""
 
 #define	CC1_ENDIAN_LITTLE_SPEC "\
@@ -1164,6 +1185,13 @@
 %{!shared:crtbegin.o%s} %{shared:crtbeginS.o%s}"
 #else
 #ifdef HAVE_LD_PIE
+#undef HARDENED_STARTFILE_SPEC
+#define	HARDENED_STARTFILE_LINUX_SPEC "\
+%{!shared: %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{static:crt1.o%s} \
+%{!static:%{nopie:crt1.o%s}%{!nopie:Scrt1.o%s}} }}} \
+%{mnewlib: ecrti.o%s} %{!mnewlib: crti.o%s} \
+%{shared:crtbeginS.o%s} %{!shared:%{static:crtbeginT.o%s} \
+%{!static:%{nopie:crtbegin.o%s}%{!nopie:crtbeginS.o%s}} }"
 #define	STARTFILE_LINUX_SPEC "\
 %{!shared: %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{static:crt1.o%s} \
 %{!static:%{pie:Scrt1.o%s}%{!pie:crt1.o%s}} }}} \
@@ -1180,6 +1208,11 @@
 #endif
 
 #ifdef HAVE_LD_PIE
+#undef HARDENED_ENDFILE_SPEC
+#define	HARDENED_ENDFILE_LINUX_SPEC "\
+%{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
+%{!static:%{nopie:crtend.o%s}%{!nopie:crtendS.o%s}} } \
+%{mnewlib: ecrtn.o%s} %{!mnewlib: crtn.o%s}"
 #define	ENDFILE_LINUX_SPEC "\
 %{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
 %{!static:%{pie:crtendS.o%s}%{!pie:crtend.o%s}} } \
@@ -1190,6 +1223,15 @@
 %{mnewlib: ecrtn.o%s} %{!mnewlib: crtn.o%s}"
 #endif
 
+#ifdef HAVE_LD_PIE
+#if defined(EFAULT_PIE) || defined(EFAULT_PIE_SSP)
+#undef STARTFILE_LINUX_SPEC
+#define STARTFILE_LINUX_SPEC HARDENED_STARTFILE_LINUX_SPEC
+#undef ENDFILE_LINUX_SPEC
+#define ENDFILE_LINUX_SPEC HARDENED_ENDFILE_LINUX_SPEC
+#endif
+#endif
+
 #define LINK_START_LINUX_SPEC ""
 
 #define LINK_OS_LINUX_SPEC "-m elf32ppclinux %{!shared: %{!static: \
--- gcc-3.3.5/gcc/config/rs6000/linux64.h.def~	Tue Nov 30 16:02:43 2004
+++ gcc-3.3.5/gcc/config/rs6000/linux64.h	Tue Nov 30 16:36:54 2004
@@ -166,6 +166,13 @@
 
 #undef  STARTFILE_LINUX_SPEC
 #ifdef HAVE_LD_PIE
+#undef HARDENED_STARTFILE_SPEC
+#undef HARDENED_STARTFILE_LINUX_SPEC
+#define HARDENED_STARTFILE_LINUX_SPEC "\
+%{!shared: %{pg|p:gcrt1.o%s} %{!pg:%{!p:%{static:crt1.o%s} \
+%{!static:%{nopie:crt1.o%s}%{!nopie:Scrt1.o%s}}}}} crti.o%s \
+%{shared:crtbeginS.o%s} %{!shared: %{static:crtbeginT.o%s} \
+%{!static:%{nopie:crtbegin.o%s} %{!nopie:crtbeginS.o%s}}}"
 #define STARTFILE_LINUX_SPEC "\
 %{!shared: %{pg|p:gcrt1.o%s} %{!pg:%{!p:%{static:crt1.o%s} \
 %{!static:%{pie:Scrt1.o%s}%{!pie:crt1.o%s}} }}} crti.o%s \
@@ -180,6 +187,11 @@
 
 #undef  ENDFILE_LINUX_SPEC
 #ifdef HAVE_LD_PIE
+#undef HARDENED_ENDFILE_SPEC
+#undef HARDENED_ENDFILE_LINUX_SPEC
+#define HARDENED_ENDFILE_LINUX_SPEC "\
+%{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
+%{!static:%{nopie:crtend.o%s}%{!nopie:crtendS.o%s}}} crtn.o%s"
 #define ENDFILE_LINUX_SPEC "\
 %{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
 %{!static:%{pie:crtendS.o%s}%{!pie:crtend.o%s}}} crtn.o%s"
@@ -188,6 +200,15 @@
 %{shared:crtendS.o%s} %{!shared:crtend.o%s} crtn.o%s"
 #endif
 
+#ifdef HAVE_LD_PIE
+#if defined(EFAULT_PIE) || defined(EFAULT_PIE_SSP)
+#undef STARTFILE_LINUX_SPEC
+#define STARTFILE_LINUX_SPEC HARDENED_STARTFILE_LINUX_SPEC
+#undef ENDFILE_LINUX_SPEC
+#define ENDFILE_LINUX_SPEC HARDENED_ENDFILE_LINUX_SPEC
+#endif
+#endif
+
 #undef  TOC_SECTION_ASM_OP
 #define TOC_SECTION_ASM_OP "\t.section\t\".toc\",\"aw\""
 
