--- gcc-3.3.5/gcc/config/sparc/linux.h.def~	Tue Nov 30 16:02:43 2004
+++ gcc-3.3.5/gcc/config/sparc/linux.h	Tue Nov 30 16:23:19 2004
@@ -40,6 +40,7 @@
    provides part of the support for getting C++ file-scope static
    object constructed before entering `main'.  */
    
+#undef  HARDENED_STARTFILE_SPEC
 #undef  STARTFILE_SPEC
 #ifdef USE_GNULIBC_1
 #define STARTFILE_SPEC \
@@ -48,6 +49,12 @@
    crti.o%s %{!shared:crtbegin.o%s} %{shared:crtbeginS.o%s}"
 #else
 #ifdef HAVE_LD_PIE
+#define HARDENED_STARTFILE_SPEC "\
+%{!shared: \
+%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{static:crt1.o%s} \
+%{!static:%{!nopie:Scrt1.o%s}%{nopie:crt1.o%s}} }}} \
+crti.o%s %{shared:crtbeginS.o%s} %{!shared:%{static:crtbeginT.o%s} \
+%{!static:%{!nopie:crtbeginS.o%s}%{nopie:crtbegin.o%s}} }"
 #define STARTFILE_SPEC "\
 %{!shared: \
 %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{static:crt1.o%s} \
@@ -69,8 +76,13 @@
    object constructed before entering `main', followed by a normal
    GNU/Linux "finalizer" file, `crtn.o'.  */
 
+#undef  HARDENED_ENDFILE_SPEC
 #undef  ENDFILE_SPEC
 #ifdef HAVE_LD_PIE
+#define HARDENED_ENDFILE_SPEC "\
+%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
+%{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
+%{!static:%{!nopie:crtendS.o%s}%{nopie:crtend.o%s}} } crtn.o%s"
 #define ENDFILE_SPEC "\
 %{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
 %{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
@@ -217,6 +229,17 @@
   "%{V} %{v:%{!V:-V}} %{!Qn:-Qy} %{n} %{T} %{Ym,*} %{Wa,*:%*} -s \
    %{fpic|fPIC|fpie|fPIE|pie:-K PIC} %(asm_cpu) %(asm_relax)"
 
+#ifdef HAVE_LD_PIE
+#if defined(EFAULT_PIE) || defined(EFAULT_PIE_SSP)
+#define HARDENED_ASM_SPEC \
+  "%{V} %{v:%{!V:-V}} %{!Qn:-Qy} %{n} %{T} %{Ym,*} %{Wa,*:%*} -s \
+   %{!D__KERNEL__: %{!fno-PIC:%{!fno-pic:%{!fno-PIE:%{!fno-pie: %{!nopie:-K PIC} }}}} } \
+   %(asm_cpu) %(asm_relax)"
+#undef ASM_SPEC
+#define ASM_SPEC HARDENED_ASM_SPEC
+#endif
+#endif
+
 /* Same as sparc.h */
 #undef DBX_REGISTER_NUMBER
 #define DBX_REGISTER_NUMBER(REGNO) (REGNO)
--- gcc-3.3.5/gcc/config/sparc/linux64.h.def~	Tue Nov 30 16:02:43 2004
+++ gcc-3.3.5/gcc/config/sparc/linux64.h	Tue Nov 30 16:27:28 2004
@@ -53,9 +53,15 @@
    provides part of the support for getting C++ file-scope static
    object constructed before entering `main'.  */
    
+#undef  HARDENED_STARTFILE_SPEC
 #undef  STARTFILE_SPEC
 
 #ifdef HAVE_LD_PIE
+#define HARDENED_STARTFILE_SPEC "\
+%{!shared:%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{static:crt1.o%s} \
+%{!static:%{!nopie:Scrt1.o%s}%{nopie:crt1.o%s}} }}} \
+crti.o%s %{shared:crtbeginS.o%s} %{!shared:%{static:crtbeginT.o%s} \
+%{!static:%{!nopie:crtbeginS.o%s}%{nopie:crtbegin.o%s}} }"
 #define STARTFILE_SPEC "\
 %{!shared:%{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} %{!p:%{static:crt1.o%s} \
 %{!static:%{pie:Scrt1.o%s}%{!pie:crt1.o%s}} }}} \
@@ -74,9 +80,14 @@
    object constructed before entering `main', followed by a normal
    GNU/Linux "finalizer" file, `crtn.o'.  */
 
+#undef  HARDENED_ENDFILE_SPEC
 #undef  ENDFILE_SPEC
 
 #ifdef HAVE_LD_PIE
+#define HARDENED_ENDFILE_SPEC "\
+%{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
+%{!static:%{!nopie:crtendS.o%s}%{nopie:crtend.o%s}} } crtn.o%s \
+%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s}"
 #define ENDFILE_SPEC "\
 %{shared:crtendS.o%s} %{!shared:%{static:crtend.o%s} \
 %{!static:%{pie:crtendS.o%s}%{!pie:crtend.o%s}} } crtn.o%s \
@@ -278,6 +289,24 @@
 %{mlittle-endian:-EL} \
 %(asm_cpu) %(asm_arch) %(asm_relax)"
 
+#ifdef HAVE_LD_PIE
+#if defined(EFAULT_PIE) || defined(EFAULT_PIE_SSP)
+#define HARDENED_ASM_SPEC "\
+%{V} \
+%{v:%{!V:-V}} \
+%{!Qn:-Qy} \
+%{n} \
+%{T} \
+%{Ym,*} \
+%{Wa,*:%*} \
+-s %{!D__KERNEL__: %{!fno-PIC:%{!fno-pic:%{!fno-PIE:%{!fno-pie: %{!nopie:-K PIC} }}}} } \
+%{mlittle-endian:-EL} \
+%(asm_cpu) %(asm_arch) %(asm_relax)"
+#undef ASM_SPEC
+#define ASM_SPEC HARDENED_ASM_SPEC
+#endif
+#endif
+
 /* Same as sparc.h */
 #undef DBX_REGISTER_NUMBER
 #define DBX_REGISTER_NUMBER(REGNO) (REGNO)
