From 372a74ad707d7c64221cf3051bbcb5f2bbdb1244 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 7 May 2017 18:28:33 -0400
Subject: [PATCH 030/112] disable COMPAT_BRK by default

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index baf796b8d924..1e15d532942b 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1876,7 +1876,6 @@ config SLUB_MEMCG_SYSFS_ON
 
 config COMPAT_BRK
 	bool "Disable heap randomization"
-	default y
 	help
 	  Randomizing heap placement makes heap exploits harder, but it
 	  also breaks ancient binaries (including anything libc5 based).
-- 
2.31.1

