--- gcc-3.4.0/gcc/config/arm/linux-elf.h.mps	Thu May 20 09:11:17 2004
+++ gcc-3.4.0/gcc/config/arm/linux-elf.h	Thu May 20 09:12:28 2004
@@ -91,10 +91,11 @@
 #define LINK_SPEC "%{h*} %{version:-v} \
    %{b} %{Wl,*:%*} \
-   %{static:-Bstatic} \
    %{shared:-shared} \
    %{symbolic:-Bsymbolic} \
+   %{!shared: %{!static: \
    %{rdynamic:-export-dynamic} \
-   %{!dynamic-linker:-dynamic-linker " ELF_DYNAMIC_LINKER "} \
+   %{!dynamic-linker:-dynamic-linker " ELF_DYNAMIC_LINKER "}} \
+   %{static:-Bstatic}} \
    -X \
    %{mbig-endian:-EB}" \
    SUBTARGET_EXTRA_LINK_SPEC
