--- gcc-3.4.3/gcc/config/sparc/linux.h.mps	Mon Nov 22 15:28:32 2004
+++ gcc-3.4.3/gcc/config/sparc/linux.h	Mon Nov 22 15:28:51 2004
@@ -177,7 +177,7 @@
 #undef ASM_SPEC
 #define ASM_SPEC \
   "%{V} %{v:%{!V:-V}} %{!Qn:-Qy} %{n} %{T} %{Ym,*} %{Wa,*:%*} -s \
-   %{fpic|fPIC|fpie|fPIE:-K PIC} %(asm_cpu) %(asm_relax)"
+   %{fpic|fPIC|fpie|fPIE|pie:-K PIC} %(asm_cpu) %(asm_relax)"
 
 /* Same as sparc.h */
 #undef DBX_REGISTER_NUMBER
--- gcc-3.4.3/gcc/config/sparc/linux64.h.mps	Mon Nov 22 15:28:40 2004
+++ gcc-3.4.3/gcc/config/sparc/linux64.h	Mon Nov 22 15:29:09 2004
@@ -281,7 +281,7 @@
 %{T} \
 %{Ym,*} \
 %{Wa,*:%*} \
--s %{fpic|fPIC|fpie|fPIE:-K PIC} \
+-s %{fpic|fPIC|fpie|fPIE|pie:-K PIC} \
 %{mlittle-endian:-EL} \
 %(asm_cpu) %(asm_arch) %(asm_relax)"
 
