From 0b5f0091fef96e5f8ee6e5da67466ca80aafbe81 Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Tue, 22 Dec 2020 23:35:53 +0100
Subject: [PATCH 043/113] stop hiding SYSFS_SYSCALL behind EXPERT

---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 9b75a4921575..006d4d41e3af 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1434,7 +1434,7 @@ config SGETMASK_SYSCALL
 	  If unsure, leave the default option here.
 
 config SYSFS_SYSCALL
-	bool "Sysfs syscall support" if EXPERT
+	bool "Sysfs syscall support"
 	default y
 	help
 	  sys_sysfs is an obsolete system call no longer supported in libc.
-- 
2.30.1

