From 2acba349ca760cbe0fdd1c24b32bfb822aef2d80 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 28 May 2017 18:51:30 -0400
Subject: [PATCH 053/113] mark kmem_cache as __ro_after_init

---
 mm/slab_common.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/slab_common.c b/mm/slab_common.c
index bff04048559f..2b73c12d8fce 100644
--- a/mm/slab_common.c
+++ b/mm/slab_common.c
@@ -33,7 +33,7 @@
 enum slab_state slab_state __ro_after_init;
 LIST_HEAD(slab_caches);
 DEFINE_MUTEX(slab_mutex);
-struct kmem_cache *kmem_cache;
+struct kmem_cache *kmem_cache __ro_after_init;
 
 #ifdef CONFIG_HARDENED_USERCOPY
 bool usercopy_fallback __ro_after_init =
-- 
2.30.1

