diff -urN gcc.old/gcc/config/alpha/linux.h gcc/gcc/config/alpha/linux.h
--- gcc.old/gcc/config/alpha/linux.h	Wed Sep  8 02:17:13 2004
+++ gcc/gcc/config/alpha/linux.h	Mon Nov  8 11:04:47 2004
@@ -36,6 +36,11 @@
 	/* The GNU C++ standard library requires this.  */	\
 	if (c_dialect_cxx ())					\
 	  builtin_define ("_GNU_SOURCE");			\
+	if (flag_pic)						\
+	  {							\
+		builtin_define ("__PIC__");			\
+		builtin_define ("__pic__");			\
+	  }							\
     } while (0)
 
 #undef LIB_SPEC
diff -urN gcc.old/gcc/config/arm/linux-elf.h gcc/gcc/config/arm/linux-elf.h
--- gcc.old/gcc/config/arm/linux-elf.h	Sat May 15 14:41:33 2004
+++ gcc/gcc/config/arm/linux-elf.h	Mon Nov  8 11:04:47 2004
@@ -89,7 +89,17 @@
    %{mbig-endian:-EB}" \
    SUBTARGET_EXTRA_LINK_SPEC
 
-#define TARGET_OS_CPP_BUILTINS() LINUX_TARGET_OS_CPP_BUILTINS()
+#define TARGET_OS_CPP_BUILTINS()		\
+  do						\
+    {						\
+	LINUX_TARGET_OS_CPP_BUILTINS();		\
+	if (flag_pic)				\
+	  {					\
+		builtin_define ("__PIC__");	\
+		builtin_define ("__pic__");	\
+	  }					\
+    }						\
+  while (0)
 
 /* This is how we tell the assembler that two symbols have the same value.  */
 #define ASM_OUTPUT_DEF(FILE, NAME1, NAME2) \
diff -urN gcc.old/gcc/config/arm/linux-gas.h gcc/gcc/config/arm/linux-gas.h
--- gcc.old/gcc/config/arm/linux-gas.h	Tue Mar 16 19:14:32 2004
+++ gcc/gcc/config/arm/linux-gas.h	Mon Nov  8 11:04:47 2004
@@ -28,7 +28,7 @@
 #define DEFAULT_SIGNED_CHAR 0
 
 #undef  SUBTARGET_CPP_SPEC
-#define SUBTARGET_CPP_SPEC  "%{posix:-D_POSIX_SOURCE} %{fPIC|fPIE:-D__PIC__ -D__pic__} %{fpic|fpie:-D__PIC__ -D__pic__}"
+#define SUBTARGET_CPP_SPEC  "%{posix:-D_POSIX_SOURCE}"
 
 #undef  SIZE_TYPE
 #define SIZE_TYPE "unsigned int"
diff -urN gcc.old/gcc/config/pa/pa-linux.h gcc/gcc/config/pa/pa-linux.h
--- gcc.old/gcc/config/pa/pa-linux.h	Sat Nov 29 04:08:11 2003
+++ gcc/gcc/config/pa/pa-linux.h	Mon Nov  8 11:04:47 2004
@@ -58,11 +58,16 @@
     {						\
 	LINUX_TARGET_OS_CPP_BUILTINS();		\
 	builtin_assert ("machine=bigendian");	\
+	if (flag_pic)				\
+	  {					\
+		builtin_define ("__PIC__");	\
+		builtin_define ("__pic__");	\
+	  }					\
     }						\
   while (0)
 
 #undef CPP_SPEC
-#define CPP_SPEC "%{fPIC|fpic|fPIE|fpie:-D__PIC__ -D__pic__} %{posix:-D_POSIX_SOURCE}"
+#define CPP_SPEC "%{posix:-D_POSIX_SOURCE}"
 
 #undef	LIB_SPEC
 #define LIB_SPEC \
diff -urN gcc.old/gcc/config/sparc/linux.h gcc/gcc/config/sparc/linux.h
--- gcc.old/gcc/config/sparc/linux.h	Wed Sep  8 02:17:19 2004
+++ gcc/gcc/config/sparc/linux.h	Mon Nov  8 11:04:47 2004
@@ -29,6 +29,11 @@
 	builtin_assert ("system=linux");	\
 	builtin_assert ("system=unix");		\
 	builtin_assert ("system=posix");	\
+	if (flag_pic)				\
+	  {					\
+		builtin_define ("__PIC__");	\
+		builtin_define ("__pic__");	\
+	  }					\
     }						\
   while (0)
 
@@ -100,7 +105,7 @@
 
 #undef CPP_SUBTARGET_SPEC
 #define CPP_SUBTARGET_SPEC \
-"%{fPIC|fPIE|fpic|fpie:-D__PIC__ -D__pic__} %{posix:-D_POSIX_SOURCE} \
+"%{posix:-D_POSIX_SOURCE} \
 %{pthread:-D_REENTRANT} %{mlong-double-128:-D__LONG_DOUBLE_128__}"
 
 #undef LIB_SPEC
diff -urN gcc.old/gcc/config/sparc/linux64.h gcc/gcc/config/sparc/linux64.h
--- gcc.old/gcc/config/sparc/linux64.h	Wed Sep  8 02:17:19 2004
+++ gcc/gcc/config/sparc/linux64.h	Mon Nov  8 11:04:47 2004
@@ -30,6 +30,11 @@
 	builtin_assert ("system=linux");	\
 	builtin_assert ("system=unix");		\
 	builtin_assert ("system=posix");	\
+	if (flag_pic)				\
+	  {					\
+		builtin_define ("__PIC__");	\
+		builtin_define ("__pic__");	\
+	  }					\
     }						\
   while (0)
 
@@ -127,7 +132,6 @@
 
 #undef CPP_SUBTARGET_SPEC
 #define CPP_SUBTARGET_SPEC "\
-%{fPIC|fpic|fPIE|fpie:-D__PIC__ -D__pic__} \
 %{posix:-D_POSIX_SOURCE} \
 %{pthread:-D_REENTRANT} \
 "
