--- gcc-3.4.4/gcc/config/alpha/elf.h.start_end~	2003-12-25 16:17:34 +0100
+++ gcc-3.4.4/gcc/config/alpha/elf.h	2005-05-23 12:12:27 +0200
@@ -399,7 +399,7 @@ void FN (void)					\
 #else
 #define STARTFILE_SPEC \
   "%{!shared: %{pg|p:gcrt1.o%s;:crt1.o%s}}\
-   crti.o%s %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+   crti.o%s %{static:crtbeginT.o%s;shared:crtbeginS.o%s;:crtbegin.o%s}"
 #endif
 
 /* Provide a ENDFILE_SPEC appropriate for ELF.  Here we tack on the
@@ -408,9 +408,15 @@ void FN (void)					\
    `main', followed by a normal ELF "finalizer" file, `crtn.o'.  */
 
 #undef	ENDFILE_SPEC
+#ifdef HAVE_LD_PIE
 #define ENDFILE_SPEC \
   "%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
    %{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s"
+#else
+#define ENDFILE_SPEC \
+  "%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
+   %{shared:crtendS.o%s;:crtend.o%s} crtn.o%s"
+#endif
 
 /* We support #pragma.  */
 #define HANDLE_SYSV_PRAGMA 1
--- gcc-3.4.4/gcc/config/arm/linux-elf.h.start_end~	2005-05-23 12:05:44 +0200
+++ gcc-3.4.4/gcc/config/arm/linux-elf.h	2005-05-23 12:12:27 +0200
@@ -80,12 +80,15 @@
    object constructed before entering `main'.  */
    
 #undef  STARTFILE_SPEC
+#ifdef HAVE_LD_PIE
 #define STARTFILE_SPEC \
-  "%{!shared: \
-     %{pg:gcrt1.o%s} %{!pg:%{p:gcrt1.o%s} \
-		       %{!p:%{profile:gcrt1.o%s} \
-			 %{!profile:crt1.o%s}}}} \
-   crti.o%s %{!shared:crtbegin.o%s} %{shared:crtbeginS.o%s}"
+  "%{!shared: %{pg|p|profile:gcrt1.o%s;pie:Scrt1.o%s;:crt1.o%s}} \
+   crti.o%s %{shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+#else
+#define STARTFILE_SPEC \
+  "%{!shared: %{pg|p|profile:gcrt1.o%s;:crt1.o%s}} \
+   crti.o%s %{shared:crtbeginS.o%s;:crtbegin.o%s}"
+#endif
 
 /* Provide a ENDFILE_SPEC appropriate for GNU/Linux.  Here we tack on
    the GNU/Linux magical crtend.o file (see crtstuff.c) which
@@ -94,8 +97,13 @@
    GNU/Linux "finalizer" file, `crtn.o'.  */
 
 #undef  ENDFILE_SPEC
+#ifdef HAVE_LD_PIE
 #define ENDFILE_SPEC \
-  "%{!shared:crtend.o%s} %{shared:crtendS.o%s} crtn.o%s"
+  "%{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s"
+#else
+#define ENDFILE_SPEC \
+  "%{shared:crtendS.o%s;:crtend.o%s} crtn.o%s"
+#endif
 
 #undef  LINK_SPEC
 #ifdef USE_UCLIBC
--- gcc-3.4.4/gcc/config/ia64/linux.h.start_end~	2005-05-23 12:05:42 +0200
+++ gcc-3.4.4/gcc/config/ia64/linux.h	2005-05-23 12:12:27 +0200
@@ -27,14 +27,20 @@ do {						\
 #else
 #define STARTFILE_SPEC \
   "%{!shared: %{pg|p|profile:gcrt1.o%s;:crt1.o%s}}\
-   crti.o%s %{shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+   crti.o%s %{shared:crtbeginS.o%s;:crtbegin.o%s}"
 #endif
 
 /* Similar to standard Linux, but adding -ffast-math support.  */
 #undef  ENDFILE_SPEC
+#ifdef HAVE_LD_PIE
 #define ENDFILE_SPEC \
   "%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
    %{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s"
+#else
+#define ENDFILE_SPEC \
+  "%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
+   %{shared:crtendS.o%s;:crtend.o%s} crtn.o%s"
+#endif
 
 /* Define this for shared library support because it isn't in the main
    linux.h file.  */
--- gcc-3.4.4/gcc/config/rs6000/linux64.h.start_end~	2005-04-29 02:31:57 +0200
+++ gcc-3.4.4/gcc/config/rs6000/linux64.h	2005-05-23 12:12:27 +0200
@@ -142,7 +142,7 @@
 #endif
 
 #define ASM_SPEC32 "-a32 %{n} %{T} %{Ym,*} %{Yd,*} \
-%{mrelocatable} %{mrelocatable-lib} %{fpic:-K PIC} %{fPIC:-K PIC} \
+%{mrelocatable} %{mrelocatable-lib} %{fpic|fPIC|fpie|fPIE:-K PIC} \
 %{memb} %{!memb: %{msdata: -memb} %{msdata=eabi: -memb}} \
 %{!mlittle: %{!mlittle-endian: %{!mbig: %{!mbig-endian: \
     %{mcall-freebsd: -mbig} \
--- gcc-3.4.4/gcc/config/rs6000/sysv4.h.start_end~	2005-05-23 12:05:43 +0200
+++ gcc-3.4.4/gcc/config/rs6000/sysv4.h	2005-05-23 12:17:15 +0200
@@ -1113,12 +1113,18 @@ extern int fixuplabelno;
 #define	STARTFILE_LINUX_SPEC "\
 %{!shared: %{pg|p:gcrt1.o%s;:crt1.o%s}} \
 %{mnewlib:ecrti.o%s;:crti.o%s} \
-%{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+%{static:crtbeginT.o%s;shared:crtbeginS.o%s;:crtbegin.o%s}"
 #endif
 
+#ifdef HAVE_LD_PIE
 #define	ENDFILE_LINUX_SPEC "\
 %{shared|pie:crtendS.o%s;:crtend.o%s} \
 %{mnewlib:ecrtn.o%s;:crtn.o%s}"
+#else
+#define	ENDFILE_LINUX_SPEC "\
+%{shared:crtendS.o%s;:crtend.o%s} \
+%{mnewlib:ecrtn.o%s;:crtn.o%s}"
+#endif
 
 #define LINK_START_LINUX_SPEC ""
 
--- gcc-3.4.4/gcc/config/sparc/linux.h.start_end~	2005-05-23 12:05:44 +0200
+++ gcc-3.4.4/gcc/config/sparc/linux.h	2005-05-23 12:12:27 +0200
@@ -61,7 +61,7 @@ Boston, MA 02111-1307, USA.  */
 #else
 #define STARTFILE_SPEC \
   "%{!shared: %{pg|p:gcrt1.o%s;:crt1.o%s}}\
-   crti.o%s %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+   crti.o%s %{static:crtbeginT.o%s;shared:crtbeginS.o%s;:crtbegin.o%s}"
 #endif
 
 /* Provide a ENDFILE_SPEC appropriate for GNU/Linux.  Here we tack on
@@ -71,9 +71,15 @@ Boston, MA 02111-1307, USA.  */
    GNU/Linux "finalizer" file, `crtn.o'.  */
 
 #undef  ENDFILE_SPEC
+#ifdef HAVE_LD_PIE
 #define ENDFILE_SPEC \
   "%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
    %{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s"
+#else
+#define ENDFILE_SPEC \
+  "%{ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
+   %{shared:crtendS.o%s;:crtend.o%s} crtn.o%s"
+#endif
 
 /* This is for -profile to use -lc_p instead of -lc.  */
 #undef	CC1_SPEC
--- gcc-3.4.4/gcc/config/sparc/linux64.h.start_end~	2005-05-23 12:05:44 +0200
+++ gcc-3.4.4/gcc/config/sparc/linux64.h	2005-05-23 12:12:27 +0200
@@ -80,7 +80,7 @@ Boston, MA 02111-1307, USA.  */
 #else
 #define STARTFILE_SPEC \
   "%{!shared:%{pg|p:gcrt1.o%s;:crt1.o%s}}\
-   crti.o%s %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbeginS.o%s}"
+   crti.o%s %{static:crtbeginT.o%s;shared:crtbeginS.o%s;:crtbeginS.o%s}"
 #endif
 
 /* Provide a ENDFILE_SPEC appropriate for GNU/Linux.  Here we tack on
@@ -90,10 +90,15 @@ Boston, MA 02111-1307, USA.  */
    GNU/Linux "finalizer" file, `crtn.o'.  */
 
 #undef  ENDFILE_SPEC
-
+#ifdef HAVE_LD_PIE
 #define ENDFILE_SPEC \
   "%{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s\
    %{ffast-math|funsafe-math-optimizations:crtfastmath.o%s}"
+#else
+#define ENDFILE_SPEC \
+  "%{shared:crtendS.o%s;:crtend.o%s} crtn.o%s\
+   %{ffast-math|funsafe-math-optimizations:crtfastmath.o%s}"
+#endif
 
 /* The GNU C++ standard library requires that these macros be defined.  */
 #undef CPLUSPLUS_CPP_SPEC
--- gcc-3.4.4/gcc/config/linux.h.start_end~	2003-11-29 04:08:10 +0100
+++ gcc-3.4.4/gcc/config/linux.h	2005-05-23 12:12:27 +0200
@@ -53,7 +53,7 @@ Boston, MA 02111-1307, USA.  */
 #else
 #define STARTFILE_SPEC \
   "%{!shared: %{pg|p|profile:gcrt1.o%s;:crt1.o%s}} \
-   crti.o%s %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+   crti.o%s %{static:crtbeginT.o%s;shared:crtbeginS.o%s;:crtbegin.o%s}"
 #endif
 
 /* Provide a ENDFILE_SPEC appropriate for GNU/Linux.  Here we tack on
@@ -63,8 +63,13 @@ Boston, MA 02111-1307, USA.  */
    GNU/Linux "finalizer" file, `crtn.o'.  */
 
 #undef	ENDFILE_SPEC
+#ifdef HAVE_LD_PIE
 #define ENDFILE_SPEC \
   "%{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s"
+#else
+#define ENDFILE_SPEC \
+  "%{shared:crtendS.o%s;:crtend.o%s} crtn.o%s"
+#endif
 
 /* This is for -profile to use -lc_p instead of -lc.  */
 #ifndef CC1_SPEC
