--- branches/gcc-4_1-branch/gcc/config/rs6000/rs6000.h	2006/04/13 20:26:59	112923
+++ branches/gcc-4_1-branch/gcc/config/rs6000/rs6000.h	2006/04/13 20:33:51	112924
@@ -1699,7 +1699,8 @@
 
 #define EASY_VECTOR_15(n) ((n) >= -16 && (n) <= 15)
 #define EASY_VECTOR_15_ADD_SELF(n) (!EASY_VECTOR_15((n))	\
-				    && EASY_VECTOR_15((n) >> 1))
+				    && EASY_VECTOR_15((n) >> 1) \
+				    && ((n) & 1) == 0)
 
 /* The macros REG_OK_FOR..._P assume that the arg is a REG rtx
    and check its validity for a certain class.
