--- branches/gcc-4_1-branch/gcc/tree-vect-transform.c	2006/05/06 17:37:53	113580
+++ branches/gcc-4_1-branch/gcc/tree-vect-transform.c	2006/05/06 17:41:48	113581
@@ -1968,6 +1968,11 @@
   then_clause = TREE_OPERAND (op, 1);
   else_clause = TREE_OPERAND (op, 2);
 
+  /* We do not handle two different vector types for the condition
+     and the values.  */
+  if (TREE_TYPE (TREE_OPERAND (cond_expr, 0)) != TREE_TYPE (vectype))
+    return false;
+
   if (!vect_is_simple_cond (cond_expr, loop_vinfo))
     return false;
 
