--- gcc-3.3.5/gcc/config/sparc/linux.h.nondef~	Fri Nov 19 08:53:00 2004
+++ gcc-3.3.5/gcc/config/sparc/linux.h	Fri Nov 19 08:54:51 2004
@@ -227,7 +227,7 @@
 #undef ASM_SPEC
 #define ASM_SPEC \
   "%{V} %{v:%{!V:-V}} %{!Qn:-Qy} %{n} %{T} %{Ym,*} %{Wa,*:%*} -s \
-   %{fpic|fPIC|fpie|fPIE:-K PIC} %(asm_cpu) %(asm_relax)"
+   %{fpic|fPIC|fpie|fPIE|pie:-K PIC} %(asm_cpu) %(asm_relax)"
 
 /* Same as sparc.h */
 #undef DBX_REGISTER_NUMBER
--- gcc-3.3.5/gcc/config/sparc/linux64.h.nondef~	Fri Nov 19 08:53:10 2004
+++ gcc-3.3.5/gcc/config/sparc/linux64.h	Fri Nov 19 08:55:03 2004
@@ -285,7 +285,7 @@
 %{T} \
 %{Ym,*} \
 %{Wa,*:%*} \
--s %{fpic|fPIC|fpie|fPIE:-K PIC} \
+-s %{fpic|fPIC|fpie|fPIE|pie:-K PIC} \
 %{mlittle-endian:-EL} \
 %(asm_cpu) %(asm_arch) %(asm_relax)"
 
