From caf7c2d41ac1f5c6b2a3561a8c4a04dbfd124597 Mon Sep 17 00:00:00 2001
From: Nick Alcock <nick.alcock@oracle.com>
Date: Mon, 9 Dec 2019 16:54:20 +0000
Subject: [PATCH 24/24] fixup! dtrace: add SDT probes

---
 net/ipv6/ip6_input.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/net/ipv6/ip6_input.c b/net/ipv6/ip6_input.c
index 259a9356fd55..ea19c5012a73 100644
--- a/net/ipv6/ip6_input.c
+++ b/net/ipv6/ip6_input.c
@@ -250,7 +250,10 @@ static struct sk_buff *ip6_rcv_core(struct sk_buff *skb, struct net_device *dev,
 	 * https://tools.ietf.org/html/draft-itojun-v6ops-v4mapped-harmful-02
 	 */
 	if (ipv6_addr_v4mapped(&hdr->saddr))
+	{
+		dropreason = "v4-mapped address in IPv6 packet";
 		goto err;
+	}
 
 	skb->transport_header = skb->network_header + sizeof(*hdr);
 	IP6CB(skb)->nhoff = offsetof(struct ipv6hdr, nexthdr);
-- 
2.28.0

