http://bugs.gentoo.org/242492
http://gcc.gnu.org/PR37408

--- gcc/gcc/function.c	2008/09/09 20:23:08	140171
+++ gcc/gcc/function.c	2008/09/09 20:24:58	140172
@@ -2336,6 +2336,11 @@
   stack_parm = gen_rtx_MEM (data->promoted_mode, stack_parm);
 
   set_mem_attributes (stack_parm, parm, 1);
+  /* set_mem_attributes could set MEM_SIZE to the passed mode's size,
+     while promoted mode's size is needed.  */
+  if (data->promoted_mode != BLKmode
+      && data->promoted_mode != DECL_MODE (parm))
+    set_mem_size (stack_parm, GEN_INT (GET_MODE_SIZE (data->promoted_mode)));
 
   boundary = data->locate.boundary;
   align = BITS_PER_UNIT;
