From 04021b5950dc86528d85249ee64ba55d9bcd009d Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Mon, 29 May 2017 07:08:42 -0400
Subject: [PATCH 020/103] set LEGACY_VSYSCALL_NONE by default

---
 arch/x86/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index d3294d58f947..7b5923dd44e1 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -2328,7 +2328,7 @@ config COMPAT_VDSO
 choice
 	prompt "vsyscall table for legacy applications"
 	depends on X86_64
-	default LEGACY_VSYSCALL_XONLY
+	default LEGACY_VSYSCALL_NONE
 	help
 	  Legacy user code that does not know how to find the vDSO expects
 	  to be able to issue three syscalls by calling fixed addresses in
-- 
2.28.0

