--- gcc-3.4.3/gcc/gcc.c.def~	Mon Nov 22 17:56:44 2004
+++ gcc-3.4.3/gcc/gcc.c	Mon Nov 22 21:18:06 2004
@@ -709,7 +718,11 @@
 
 static const char *asm_debug;
 static const char *cpp_spec = CPP_SPEC BOUNDS_CPP_SPEC BC_STRINGS_ONLY_CPP_SPEC;
+#ifdef HARDENED_CC1_SPEC_SUFFIX
+static const char *cc1_spec = CC1_SPEC " " HARDENED_CC1_SPEC_SUFFIX;
+#else
 static const char *cc1_spec = CC1_SPEC;
+#endif
 static const char *cc1plus_spec = CC1PLUS_SPEC;
 static const char *link_gcc_c_sequence_spec = LINK_GCC_C_SEQUENCE_SPEC;
 static const char *asm_spec = ASM_SPEC;
